{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 5 Lab: The dark matter of the genome (ChIP-seq)\n",
    "\n",
    "Skills: ChIP-seq, genome browsers, peak calling, motif finding\n",
    "\n",
    "## Intro\n",
    "In 2006, there was a striking discovery that if you treat adult cells with a specific set of transcription factors, they could become \"pluripotent\", meaning they can then be programmed to theoretically any type of cell. These four transcription factors (Oct4/*Pou5f1*, Sox2, cMyc, and Klf4) are collectively known as \"Yamanaka factors\" (named after Shinya Yamanaka's lab, which originally showed this).\n",
    "\n",
    "In this lab, we will analyze ChIP-sequencing from three of these factors (Oct4, Sox2, and Klf4) as well as two histone modifications (H3K4me2 and H3K27ac, which often mark regulatory/enhancer regions) in mouse embryonic stem cells, which are pluripotent and express the Yamanaka factors. We'll use the ChIP-seq data to determine where these factors and modifications are binding and which specific sequence motifs the factors are binding to.\n",
    "\n",
    "In this lab, we'll go through:\n",
    "1. Aligning ChIP-seq reads to a referenc genome\n",
    "2. Visualizing ChIP-seq data in IGV.\n",
    "3. Identifying binding sites from ChIP-seq data (\"peak calling\")\n",
    "4. Visualizing patterns of transcription factor and histone modification binding.\n",
    "5. Motif finding to identify the sequence a transcription factor binds to.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Intro-Q1)**</font> In your own words, summarize background and motivation for the study. See the report notebook for more suggestions on how to write this.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Intro-Q2)**</font> In your own words, summarize your main results. See the report notebook for more suggestions on how to write this. You should probably wait until you've actually completed the analyses to write this part though!\n",
    "\n",
    "\n",
    "### Summary of tools covered\n",
    "In this lab we'll be using the following tools:\n",
    "\n",
    "* [BWA MEM](https://github.com/lh3/bwa): for aligning our reads to the reference genome\n",
    "* [IGV](https://software.broadinstitute.org/software/igv/): a genome browser. Used for visualizing ChIP-seq alignments and peaks (and other types of genomic data)\n",
    "* [HOMER](http://homer.ucsd.edu/homer/): a toolkit that supports a large number of analyses of ChIP-seq datasets (and was created right here at UCSD by Chris Benner!) These include:\n",
    "  * Peak calling (identifying where the binding sites are from a ChIP-seq dataset)\n",
    "  * Annotating peaks and generating coverage histogram plots\n",
    "  * Motif finding (finding which sequences our transcription factor is binding to).\n",
    "\n",
    "In the last part, you may also use:\n",
    "* [MEME](http://meme-suite.org/): a suite of tools for motif analysis. Specifically, you'll probably use the `fimo` tool that is part of this package for scanning a genome for motif matches.\n",
    "* [BEDtools](https://bedtools.readthedocs.io/en/latest/) a set of tools for wrangling BED files and other genomic data formats.\n",
    "\n",
    "Note there are many tools available for performing these tasks. For a list of additional relevant tools see the lecture slides.\n",
    "\n",
    "And as usual, we'll do some plotting examples with the [matplotlib](https://matplotlib.org/) Python library, but you can use whatever you're most comfortable with to plot. \n",
    "\n",
    "\n",
    "### Summary of data provided\n",
    "Data for this week can be found in `/datasets/cs185-sp21-A00-public/week5`. You should see:\n",
    "\n",
    "* `Oct4.esc.fastq`: reads from ChIP-seq of the transcription factor Oct4.\n",
    "* `Klf4.esc.fastq`: reads from ChIP-seq of the transcription factor Klf4.\n",
    "* `Sox2.esc.fastq`: reads from ChIP-seq of the transcription factor Sox2.\n",
    "* `H3K27ac.esc.fastq`: reads from ChIP-seq of the histone modification H3K27ac.\n",
    "* `H3K4me2.esc.fastq`: reads from ChIP-seq of the histone modification H3K4me2.\n",
    "* `input.esc.fastq`: reads from an input control (whole cell extract) for the ChIP-seq. Recall this comes from performing ChIP-seq but without the antibody pull down step.\n",
    "\n",
    "For the motif analysis in the last part, you'll also see:\n",
    "* `GRCm38.chr17.fa`: chr17 extracted from the GRCm38 reference genome.\n",
    "* `Oct4_motif.meme`: the MEME-format motif for Oct4.\n",
    "\n",
    "Recall the mouse reference genome can be found at: `/datasets/cs185-sp21-A00-public/genomes/GRCm38.fa` (build GRCm38). For this week, we have also added the corresponding bwa index files in the same directory.\n",
    "\n",
    "**Acknowledgements**: Parts of this lab are modified from material originally written by Chris Benner and revised by Alon Goren.\n",
    "\n",
    "## Tips\n",
    "\n",
    "You may include an image file (such as an IGV screenshot) by:\n",
    "* saving the image on your local computer (png files should work)\n",
    "* navigating in JupyterHub to the assignment directory.\n",
    "* uploading the file\n",
    "* Using HTML (e.g. `<img src=filename>`) to include the file in a markdown cell\n",
    "\n",
    "Be aware of chromosome name mismatches. e.g. \"chr17\" vs. \"17\".\n",
    "\n",
    "For long running commands, `nohup` is your friend."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Aligning ChIP-seq reads\n",
    "\n",
    "Before getting started, it is always a good idea to get some basic stats on the data you're dealing with.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q1)**</font> Summarize the datasets we are starting with: what samples are we analyzing? Where did they come from? How many reads are there in each dataset, and what is the read length?\n",
    "\n",
    "As in our previous NGS analyses, the first step here will be to align the reads to a reference genome. We recommend using `bwa mem`. Type the command at the terminal to see usage, or look back to Lab 1 to recall the syntax for running BWA. The genome at `/datasets/cs185-sp21-A00-public/genomes/GRCm38.fa` has already been indexed using `bwa index`.\n",
    "\n",
    "Align reads from each dataset to the GRCm38 (mm10) reference genome. You may want to use UNIX for loops to avoid retyping the command many times (see example for loop below for review). You may also want to use multithreading (e.g. `-t 6` to speed this up, but make sure you're on the high memory instances if you do that.)\n",
    "\n",
    "```shell\n",
    "# Example for loop syntax. You will have to modify paths this to actually work on your data\n",
    "for prefix in Oct4 Klf4 Sox2\n",
    "\n",
    "do\n",
    "    bwa mem GRCm38.fa ${prefix}.fastq | samtools view -bS > ${prefix}.bam\n",
    "done\n",
    "\n",
    "```\n",
    "\n",
    "Store the results in `~/week5`. After aligning, use `samtools` to sort and index the resulting BAM file (again, look back to Lab 1 if you need a refresher how to do this).\n",
    "\n",
    "After alignment finishes, it is a good practice to take a look at the resulting BAM files using `samtools view`. e.g.:\n",
    "\n",
    "```shell\n",
    "samtools view mysample.bam | less -S\n",
    "```\n",
    "\n",
    "You can also see how many reads aligned to each chromosome using e.g.:\n",
    "```\n",
    "samtools view ~/week5/bams/input.sorted.bam| cut -f 3 | uniq -c\n",
    "```\n",
    "\n",
    "(You should see that most reads were aligned to chr17, since we extracted only those reads to make this analysis run faster. But there will be reads aligned to other chromosomes as well which are mostly mapping errors).\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q2)**</font> Summarize the methods you used to align the data. What aligner and version did you use? What build of the reference genome?\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q1)**</font> What percentage of your reads from each dataset were successfully aligned? Recall `samtools flagstat` can be used to easily find this number.\n",
    "\n",
    "(Alignment might take a couple minutes. In the mean time, you might want to get started installing and exploring IGV in part 3)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Getting started with HOMER (making tag directories)\n",
    "\n",
    "We will be using Homer for most of the analyses in this lab. For most analyses, Homer requires us to perform a preprocessing step to convert the BAM files into \"tag directories\". Tag directories are analogous to sorted BAM files and are the starting point for most HOMER operations like finding peaks, creating visualization files, or calculating read densities. The command also performs several quality control and parameter estimation calculations.\n",
    "\n",
    "The command `makeTagDirectory` will do this for us. The syntax is:\n",
    "\n",
    "```shell\n",
    "makeTagDirectory <output directory> <input BAM file> [options]\n",
    "```\n",
    "\n",
    "For example, if you made a BAM file `~/week5/bams/Oct4.sorted.bam`, you might run:\n",
    "\n",
    "```shell\n",
    "makeTagDirectory ~/week5/tagdirs/Oct4 ~/week5/bams/Oct4.sorted.bam\n",
    "```\n",
    "\n",
    "This command is going through the BAM file and doing lots of preprocessing steps: removing reads that do not align to a unique position in the genome, separating reads by chromosome and sorting them by position, calculating how often reads appear in the same position to estimate the clonality (i.e. PCR duplication), calculating the relative distribution of reads relative to one another to estimate the ChIP-fragment length, calculating sequence properties and GC-content of the reads and performing a simple enrichment calculation to check if the experiment looks like a ChIP-seq experiment (vs. an RNA-seq experiment).\n",
    "\n",
    "The command creates a new directory, in this case named `~/week5/tagdirs/Oct4`. Inside the directory are several test files that contain various QC results. Don't worry too much about those, but you can try looking at the output files by typing `less -S <filename>` to see what's there. \n",
    "\n",
    "Run this on all of your aligned BAM files. It shouldn't take long (seconds) for each sample to finish."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Visualizing the data with IGV\n",
    "\n",
    "Next we will visualize the ChIP-seq experiments by creating [bedGraph files](https://genome.ucsc.edu/goldenPath/help/bedgraph.html) from the tag directories and using the IGV genome browser to look at the results. BedGraph files are similar to BED files we've seen in the past (with chrom, start, and end columns) except now with a 4th column giving a datavalue. These files are useful for describing how many reads aligned to each region of the genome.\n",
    "\n",
    "We will create these files using the `makeUCSCfile` command that is part of Homer. For most ChIP-seq experiments all you need to do is specify the tag directory and specify `-o auto` for the command to automatically save the bedGraph file inside the tag directory. e.g.:\n",
    "\n",
    "```shell\n",
    "makeUCSCfile ~/week5/tagdirs/Oct4 -o auto\n",
    "```\n",
    "\n",
    "This creates the file `~/week5/tagdirs/Oct4/oct4-esc.ucsc.bedGraph.gz`. This file format specifies the normalized read depth at variable intervals along the genome (use `zcat` and the filename to view the file format for yourself).\n",
    "\n",
    "Use the `makeUCSCfile` command to make bedGraph files for each of your samples. \n",
    "\n",
    "Now, to visualize with IGV, first download the bedGraph files to your local computer. You can download files from datahub by navigating to the directory where the file is stored, clicking the box next to it, and clicking \"download\".\n",
    "\n",
    "Then, go to IGV. There is a [desktop version](https://software.broadinstitute.org/software/igv/download) and a javascript version (https://igv.org/app/). We *highly* recommend the desktop version, which can be easier to use and faster. Instructions below are the mostly the same regardless of which one you use.\n",
    "\n",
    "Make sure to select the \"mm10\" genome (which is equivalent to GRCm38). Use \"Tracks->Local file\" to upload your bedgraph files.\n",
    "\n",
    "The tracks will display the relative density (coverage) of ChIP-seq reads at each position in the genome. Zoom in chr17, since that is where all of data is from. \n",
    "\n",
    "See if there are any interesting patterns in the data that catch your eye. Try visiting the *Pou5f1* locus (the gene for Oct4) by typing the gene name into the search bar at the top. Once at the *Pou5f1* locus, zoom out to see if there any nearby sites with possible binding sites in each of your tracks.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q2)**</font> Based on visual inspection of the datasets on IGV, where are the peaks for each of the datasets falling (promoters? gene regions? exons? elsewhere?). How do the signals for the transcripton factors (Oct4, Sox2, and Klf4) compare to the signals for the two histone modifications? What does the input signal look like? You may wish to include a screenshot of the IGV visualization in your report. The lecture slides provide some hints for the types of patterns you should be seeing for each dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Identifying binding sites (peaks)\n",
    "\n",
    "One of the most common tasks with ChIP-seq data is to find regions that enriched (compared to a control like whole cell extract). These enriched regions are commonly called \"peaks\". We will use the `findPeaks` utility from Homer, which takes tag directories as input and outputs a set of peak regions.\n",
    "\n",
    "The general syntax of the command is:\n",
    "\n",
    "```shell\n",
    "findPeaks <tag directory> -i <control tag directory> -style XX -o auto\n",
    "```\n",
    "\n",
    "for example:\n",
    "```shell\n",
    "findPeaks ~/week5/tagdirs/Oct4 -i ~/week5/tagdirs/input -style factor -o auto\n",
    "```\n",
    "\n",
    "where we will use `-style factor` for transcription factors and `style histone` for histone modifications. Generally, for transcription factors we will be looking for more focal, narrower peaks whereas for histone modifications we will be looking for broader peaks. This flag tells Homer which type of peaks to look for.\n",
    "\n",
    "This command will look for enriched regions and filter them based on several criteria, including ensuring that they have at least 4-fold more reads in peak regions relative to the control experiment (in this case `input`). The output will be stored in a HOMER-style peak file located in the Oct4 tag directory (`Oct4/peaks.txt`). This file will be called `regions.txt`, rather than `peaks.txt`, for histone modifications. The beginning of this file contains statistics and QC stats from the peak finding, including the number of peaks, number of peaks lost to input filtering, etc. \n",
    "\n",
    "One field worth paying attention to is the \"Approximate IP efficiency\" which reports what fraction of reads from the experiment were actually found in peaks. For most decent experiments this value ranges from 1% to >30% (remember ChIP is an enrichment strategy... there is plenty of background in the data too!). \n",
    "\n",
    "After the header lines (which begin with `#`, you'll see the actual peaks, with one line per peak. There are a lot of columns. The important ones for us are: columns 2-4 give the chromosome start end, column 11 gives the fold change over the background, and column 12 gives the p-value.\n",
    "\n",
    "Use `findPeaks` to call peaks in each sample. Be sure to set `style` appropriately for each dataset.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q3)**</font> Summarize the methods you used to call peaks for each sample the data. What tool did you use? Which version? Did you use any non-default parameters? Did you use a different command for transcription factors vs. histone modifications?\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q3)**</font> Summarize peak-finding results: what was the IP efficiency reported for each dataset? How many peaks did you find? What was the average peak size for each dataset? Note you should only have 5 peak sets, since you won't have peaks for the input control.\n",
    "\n",
    "One other thing to note is that HOMER reports the results in a \"peak\" file, which has a slightly different format from a traditional BED file format. To create a BED file from the peak file, use the Homer utility `pos2bed.pl` following the sample below. BED files can be uploaded to IGV just like a bedGraph file. Also, most HOMER programs will work with either BED or peak files as input. Example command:\n",
    "```\n",
    "pos2bed.pl Oct4/peaks.txt > Oct4/Oct4.peaks.bed`\n",
    "```\n",
    "\n",
    "Copy the peaks (in BED format) you called to your local computer, and upload the resulting files to IGV. Explore how they look on IGV: what does the coverage profile look like for peaks with high scores? low scores?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Visualizing binding patterns\n",
    "\n",
    "We will next use the Homer utility `annotatePeaks.pl` to learn about where our peaks are falling. This tool can do a lot of different things. For now we'll use it to visualize read counts around promoter regions.\n",
    "\n",
    "The general usage of `annotatePeaks.pl` is:\n",
    "```shell\n",
    "annotatePeaks.pl <peaks file> <genome fasta> [options] > output.txt\n",
    "```\n",
    "\n",
    "We can use the `-hist` option to `annotatePeaks.pl`, which gives the relative count of reads for each dataset centered around a specific feature. Here, we will look at read counts relative to transcription start sites (TSSs), which mark the center of promoter regions. The command below can be used to create histogram data for a list of tag directories. Note you'll have to change this command if your tag directories had different names.\n",
    "\n",
    "```shell\n",
    "annotatePeaks.pl tss \\\n",
    "  /datasets/cs185-sp21-A00-public/genomes/GRCm38.fa \\\n",
    "  -size 8000 \\\n",
    "  -hist 10 \\\n",
    "  -d ~/week5/tagdirs/Oct4 ~/week5/tagdirs/Sox2 ~/week5/tagdirs/Klf4 ~/week5/tagdirs/H3K4me2 ~/week5/tagdirs/H3K27ac \\\n",
    "  -gtf /datasets/cs185-sp21-A00-public/genomes/GRCm38.75.gtf > ~/week5/annotations/tss_histogram.txt\n",
    "```\n",
    "\n",
    "Here is a brief description of the options used:\n",
    "* `tss` is a built in shortcut telling Homer to look at regions centered on transcription start sites. We could have also provided a peaks file with regions to plot or a BED file listing locations of transcription start sites.\n",
    "* `GRCm38.fa` tells it where the reference genome is\n",
    "* `-size 8000` tells it look at 8000bp regions centered at each TSS\n",
    "* `-hist 10` gives the bin size for our histograms\n",
    "* `-d tagdir1 tagdir2...` gives a list of tag directories\n",
    "* `-gtf` points it to the gene annotations, which it needs to figure out where the TSSs are.\n",
    "\n",
    "Take a look at the output file. You’ll notice that the first column gives the distance offsets from the TSS followed by columns corresponding to the ‘coverage’, ‘+ Tags’, and ‘- Tags’ for each experiment. The code in the cell below shows how to load this histogram to a pandas dataframe and plot the distribution of Oct4 and H2K27ac around TSSs. You may hear these types of plots referred to as \"composite plots\".\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q4)**</font> Summarize the methods you used to annotate peaks for each sample the data and visualize read counts around transcription start sites. What tool did you use? Which version? \n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q4)**</font>: Generate a composite plot showing the distribution of all five datasets around TSSs. Be sure to label your axes and provide a legend or description of which color denotes which dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dist.from.tss</th>\n",
       "      <th>Oct4</th>\n",
       "      <th>Sox2</th>\n",
       "      <th>Klf4</th>\n",
       "      <th>H3K4me2</th>\n",
       "      <th>H3K27ac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-4000</td>\n",
       "      <td>0.894853</td>\n",
       "      <td>0.816687</td>\n",
       "      <td>0.929456</td>\n",
       "      <td>1.465848</td>\n",
       "      <td>1.165416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-3990</td>\n",
       "      <td>0.890395</td>\n",
       "      <td>0.807531</td>\n",
       "      <td>0.927906</td>\n",
       "      <td>1.467393</td>\n",
       "      <td>1.170515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-3980</td>\n",
       "      <td>0.881561</td>\n",
       "      <td>0.804261</td>\n",
       "      <td>0.920969</td>\n",
       "      <td>1.469015</td>\n",
       "      <td>1.171071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-3970</td>\n",
       "      <td>0.888774</td>\n",
       "      <td>0.802299</td>\n",
       "      <td>0.920071</td>\n",
       "      <td>1.469324</td>\n",
       "      <td>1.171164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-3960</td>\n",
       "      <td>0.882209</td>\n",
       "      <td>0.792080</td>\n",
       "      <td>0.915256</td>\n",
       "      <td>1.482844</td>\n",
       "      <td>1.179044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   dist.from.tss      Oct4      Sox2      Klf4   H3K4me2   H3K27ac\n",
       "0          -4000  0.894853  0.816687  0.929456  1.465848  1.165416\n",
       "1          -3990  0.890395  0.807531  0.927906  1.467393  1.170515\n",
       "2          -3980  0.881561  0.804261  0.920969  1.469015  1.171071\n",
       "3          -3970  0.888774  0.802299  0.920071  1.469324  1.171164\n",
       "4          -3960  0.882209  0.792080  0.915256  1.482844  1.179044"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/matplotlib/cbook/__init__.py:1402: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.\n",
      "  x[:, None]\n",
      "/opt/conda/lib/python3.6/site-packages/matplotlib/axes/_base.py:276: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.\n",
      "  x = x[:, np.newaxis]\n",
      "/opt/conda/lib/python3.6/site-packages/matplotlib/axes/_base.py:278: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.\n",
      "  y = y[:, np.newaxis]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbAAAADQCAYAAABmzTTEAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3dd5yU1dXA8d/Z3htbKAvs0jsoRRBFFIMNe0ys0Rj1tcVuyquJmmJiNMaWBI2vxt7RSOxIUZEiHaQtnaXtsr3v7Mx9/7jPwi5smYWdnRn2fD+f+czMU88us5y597nPuWKMQSmllAo2If4OQCmllDoSmsCUUkoFJU1gSimlgpImMKWUUkFJE5hSSqmgpAlMKaVUUArzdwANpaammqysLH+HoZRSKkAsXbp0vzEmral1AZXAsrKyWLJkib/DUEopFSBEZHtz67QLUSmlVFDSBKaUUiooaQJTSikVlALqGlhTXC4Xubm5VFdX+zuUoBMVFUVmZibh4eH+DkUppdpdwCew3Nxc4uPjycrKQkT8HU7QMMZQUFBAbm4u2dnZ/g5HtZfyfFj8LIy7AeLS/R2NUn4V8F2I1dXVdOnSRZNXG4kIXbp00ZbrscQY+OI38NWj8OXv/B2NUn4X8AkM0OR1hPT3dozZtRRWvmFf56/3byxKBYCgSGD+lpuby/nnn0///v3p27cvt99+O7W1tc1u/8QTT1BZWXnY8vPOO49hw4b5MlR1LNs6zz4PuQD259gWmVKdmCawVhhjuOiii7jgggvIyclh48aNlJeXc9999zW7T1MJbMaMGcTFxfk6XHUs2/4tpA2GnuOguhiqivwdkVJ+pQmsFbNnzyYqKoqf/vSnAISGhvK3v/2NF154gYqKCu655x6GDx/OiBEjePrpp3nqqafYvXs3p556KqeeeioA5eXlPP7449x///3+/FFUMPO4Ycci6D0BEjPtspJc/8aklJ8F/CjEhh6a+T1rd5e26zGHdE/ggXOHNrv++++/Z/To0Y2WJSQk0KtXL55//nm2bt3K8uXLCQsLo7CwkJSUFB5//HHmzJlDamoqAL/5zW+4++67iYmJadfYVSeydzXUlkGvEw8msNJd0G2Ef+NSyo+0BdYKY0yTgyGMMXz11VfceOONhIXZ7wEpKSmHbbdixQo2bdrEhRde6PNY1TFsxwL73HsCJGgLTCkIshZYSy0lXxk6dCjvvfdeo2WlpaXs3LmTPn36tDrSb8GCBSxdupSsrCzq6urIy8tj8uTJzJ0714dRq2PO9vmQ1Mu2vjweCAmzLTClOjFtgbViypQpVFZW8vLLLwPgdru5++67ueaaa5g6dSrTp0+nrq4OgMLCQgDi4+MpKysD4KabbmL37t1s27aNb775hgEDBmjyUm1jDGxfYLsPAUJCIDYNKvL9G5dSfqYJrBUiwvvvv88777xD//79GTBgAFFRUTz88MNcd9119OrVixEjRjBy5Ehef/11AG644QbOOuusA4M4lDoqBZugcj/0PvHgstg0W5VDqU5MTADdSzJmzBhz6Hxg69atY/DgwX6KKPjp7+8YsPQlmHkb3LoEUvvbZa9cZIfR3zDHv7Ep5WMistQYM6apddoCUyrQ7fseIuKgS7+Dy+LStQtRdXqawJQKdPnrIXUANBwwFJtqE1gA9aAo1dE0gSkV6Pbn2ATWUGw61FVDTZl/YlIqAGgCUyqQuV1QtgeSezdeHptmn7UbUXVimsCUCmSluwEDCT0aL4/TBKaUJjClAln9zcqJhySwWGcyS01gqhPTBOaFQ6vI//vf/+bWW28FYPr06QwfPpxRo0Zx0kknsXbtWgDmzp3LtGnTDuxz//33c8YZZ1BTU8MVV1zBwIEDGTZsGNdeey0ulwuARx99lFGjRjFq1CiGDRtGaGjogZujVSdVXy4qsWfj5fVdiOV5HRuPUgFEE9hRuvzyy1m9ejUrVqzgF7/4BXfddddh2/zxj39k/vz5fPDBB0RGRnLFFVewfv16Vq9eTVVVFc8//zwA9957LytWrGDFihX86U9/4pRTTmmyvqLqROoT2KFdiLG2ULS2wFRnpgnsKCUkJBx4XVFRcVhtxL/+9a98/PHHzJw5k+joaADOPvtsRAQRYdy4ceTmHl6U9Y033uCyyy478P6CCy5g9OjRDB06lOeee+7A8k8//ZTjjz+ekSNHMmXKlPb+8ZS/leRCVBJEHjKXXGg4RCdrAlOdms+L+YpIKLAE2GWMmdba9i365Fd2Won21HU4nPXnFjepqqpi1KhRB94XFhZy3nnnHXj/97//nccff5za2lpmz559YPn8+fPZsGEDS5cubXIyS5fLxSuvvMKTTz7ZaHllZSWffvopzzzzzIFlL7zwAikpKVRVVTF27FguvvhiPB4P119/PV999RXZ2dna3XgsKt11cPqUQ8Wmaxei6tQ6ogV2O7CuA87jM9HR0Qe69lasWMHvfve7RutvueUWNm/ezCOPPMIf/vCHA8v79euHMYbPP/+8yePefPPNTJo0iZNPPrnR8pkzZzJx4sRG3YdPPfUUI0eOZPz48ezcuZOcnBwWLlzIpEmTyM7OBpqezkUFubK9EN+16XWxaVCxv2PjUSqA+LQFJiKZwDnAH4HDLw61VSstJX+79NJLuemmmw68z8jI4LXXXmPKlCl06dKlUXHfhx56iPz8fJ599tnDjvPmm2826j6cO3cus2bNYsGCBcTExDB58mSqq6ubnatMHUMqCyFtYNPr4tLav0dCqSDi6xbYE8AvAI+Pz+M3OTk5B15/9NFH9O/fv9H6AQMGMGPGDK688kpWrFgBwPPPP89nn33GG2+8QUhI43+CkpIS5s2bx/nnn99oWXJyMjExMaxfv56FCxcCMGHCBObNm8fWrVsBtAvxWFRVCNHNtKx1ShXVyfmsBSYi04A8Y8xSEZncwnY3ADcA9OrVy1fh+MwzzzzDrFmzCA8PJzk5mZdeeumwbcaOHcuLL77Ieeedx5w5c7jxxhvp3bs3EyZMAOCiiy7it7/9LQDvv/8+U6dOJTY29sD+Z555JtOnT2fEiBEMHDiQ8ePHA5CWlsZzzz3HRRddhMfjIT09nS+++KIDfmrVIepqobYcYro0vT42HapLoK4GwiI7NjalAoDPplMRkT8BVwF1QBSQAMwwxlzZ3D46nUr7099fECvbC38dCOf8FcZed/j6JS/Cf++AO9cefqOzUscIv0ynYoz5tTEm0xiTBVwKzG4peSmlDlHpdAk314UYV1+NQ0ciqs6p1QQmIo94s0wp1c4qC+xzTHPXwJwEpjMzq07KmxbYD5pYdlZbTmKMmXvU94Ap1dlUtdIC02ocqpNrdhCHiNwE3Az0EZFVDVbFA/N9HVhDOlz8yPjq+qbqIPVdiM21wLQLUXVyLY1CfB34BPgT8KsGy8uMMR02XjsqKoqCggK6dOmiSawNjDEUFBQQFRXl71DUkWqtBRYRC+ExejOz6rSaTWDGmBKgBLjMKQeV4WwfJyJxxpgdHRFgZmYmubm55OdrN0lbRUVFkZnZTBkiFfgqCyEsGiJimt8mNk3LSalOq9X7wETkVuBBYB8Hb0g2wAjfhXVQeHj4gVJJSnUqVUXNdx/W05uZVSfmzY3MdwADjTEFvg5GKdVAdQlEJba8TVw6FHdIZ4hSAcebUYg7sV2JSqmOVFthr3O1JDZVW2Cq0/KmBbYFmCsiHwE19QuNMY/7LCqllJcJLN0O4vB4IESn91OdizcJbIfziHAeSqmOUFthr3G1JC4djNuOWKy/L0ypTqLVBGaMeagjAlFKHcLlTQvMSXDl+zSBqU7Hm1GIc7CjDhsxxpzmk4iUUpY3XYhJzgwOxTshY6jvY1IqgHjThXhPg9dRwMXYCvNKKV/yJoEl9rTPJTt9H49SAcabLsSlhyyaLyLzfBSPUgrsoAxXZesJLC4dwqKgeHvHxKVUAPGmC7HhnZQhwGigq88iUkrZ5AWtJzAR2wrTe8FUJ+RNF+JS7DUwwXYdbgV+5suglOr0aivsc3gLZaTqJfWy18CU6mS86ULUOk5KdTSXk8Ai4lrfNqkn7Fnp23iUCkDeTGgZLiK3ici7zuNWEQnviOCUCnquKlj1Nuxe0bb96ltgrXUhgm2BVe6HmrK2x6dUEPOmC/GfQDjwD+f9Vc6y63wVlFLHjFd/CNu/sa/vyTk4h1drauuvgXnRhZg+xD7nrYOe49oeo1JBypvaM2ONMVcbY2Y7j58CY30dmFJBr3jnweQF8OFt3u9bW26fvelC7DrcPre1ladUkPMmgblFpG/9GxHpA7h9F5JSx4jNX9rnWxbDSXfCxk+97+ZrSxdiQg9IGwSLnwW368hiVSoIeZPA7gXmiMhc5/6v2cDdvg1LqWPAlnkQ1xVSB0DviYDxvpVUP4zem1GIIjDlASjYBN9/cMThKhVsvBmF+KWI9AcGYofSrzfG1LSym1Kdm8cDW7+CfqfbBNP9eLt89zLIPrn1/dvShQgw4Ew7d9jiZ6HXeDsyUaljnDejEG8Boo0xq4wxK4EYEbnZ96EpFcRyPrMjAwdMte9ju9jRgruWebd/W7oQwU6lknUy5H4HTwyD7d+2PWalgow3XYjXG2OK698YY4qA630XklLHgCUv2GtTg88/uKzHaNsC80ZtG7oQ6017AkZeZl+/+kOoLPR+X6WCkDcJLEREpP6NiISi84Ip1bzyPNj0JYz4MYQ26KXvNsqWfPImsdSW2+TVlkkq49Lgwulw43x7I/TcP7U9dqWCiDd/HZ8Bb4vIFBE5DXgD+NS3YSkVpOpq4d1r7SSTIy9tvK7bSPvsTdWM2oq2tb4a6joMhl0M3/0f1OnlanXs8iaB/RL4ErgJuMV5/QtfBqVU0CneCSteh+dOgW1fw9mPQdrAxtvUvy/c0vrxvKlE35KBZ9skun/jkR9DqQDnzShEDzDdeSilDrV9AbxyIdRV2ffjboBxTVwmjusKoZFQtK31Y9ZWeD8CsSkZw+zznpUHb3RW6hjThg52pVSTFk23JZ+u/RxOux9O/d+mtwsJgeTeULS19WPWlntXRqo5qQMgLgM2fHLkx1AqwGkCU+po7F4B6z60o/96nQCT7oXo5Oa3T87ysgV2lF2IISE2pvX/ha1fH/lxlApgmsCUOlIVBfDm5RDTxSYubyRnQdF2MKbl7Y62CxFg8q9tK2yR9v6rY1Oz18BEZCZ2IssmGWPO80lESgWLNe9C6S74yX8gOsm7fZKzoKYUqoogJqX57eqH0R+N8CjoO8XWZDTGVgRR6hjSUgvsMeCv2BmYq4B/OY9yYE1rBxaRKBFZLCIrReR7EXmoPQJWKmB8/wGkD4U+k73fJ9mZH7a162BHOwqxXvfjoHyfvf9MqWNMswnMGDPPGDMPOM4Y82NjzEzncTlwkhfHrgFOM8aMBEYBZ4rI+PYJWyk/K8+HHQtg8Llt2y85yz63dh2stqJ9Elj/H9jnVW8f/bGUCjDeXANLc6ZQAUBEsoG01nYyllORlHDn0UrHv1JBYsPHgIHB09q2X3Jv+1zYQgvM42m/FlhKNmSOhZzPj/5YSgUYbxLYncBcZzqVucAc4A5vDi4ioSKyAsgDvjDGLGpimxtEZImILMnPz29D6Er50bqZtjVVf7+VtyJiITa95RZY/VQq7ZHAwBb53b0Maspb31apINJqAjPGfAr0B253HgONMZ95c3BjjNsYMwrIBMaJyGF/7caY54wxY4wxY9LSWm3YKeV/FQWwZa7tPjySgRGtDaVvayX61mSfDJ462LGwfY6nVIDwdhh9/XxgI4Efi8hP2nISp5r9XODMNkWnVCBa9RZ4XDDy8iPbv34ofXPaOhdYa3qOh5Bw2PZV+xxPqQDhzXxgDwBPO49Tgb8ArQ6hF5E0EUlyXkcDpwPrjypapfytaDvMe8ROjZIx5MiOkZwFpbm28G9T2rsFFhFjr4Nt1QSmji3etMB+CEwB9hpjfopthUV6sV83YI6IrAK+w14D++8RR6pUIPjuX/Y+rnMeP/JjJPUE44HyvU2vr09gR3sfWEP9psDu5bB/U/sdUyk/8yaBVTkFfetEJAE7IKNPK/vgzOB8nDFmhDFmmDHmd0cbrFJ+5XbByrdspffuo478OLHp9rk8r+n1rvoWWDt1IQKMutwmxI/ubL9jKuVn3iSwJU5X4L+ApcAyYLFPo1Iq0BgDsx6Eijw4vk2XgA8X10oCa+8uRICE7jDmWls5X+cIU8cIb0Yh3myMKTbGTAd+AFztdCUq1XnsWgoLnoE+p0L/qUd3rLgM+1y+r+n1vkhgAJlj7OCTdTPb97hK+Yk3gzhERK4Ukd8aY7YBxSIyzvehKRUgPG74z60QHguX/PvoawrGOreLtNoCa8cuRLCJNzkbvnmifY+rlJ9404X4D2ACcJnzvgz4u88iUirQrH4H8tfBOX/1vmhvS8Ii7JQrFa0lsHYcxAG2RTf6ati3Gkp3t++xlfIDbxLYCcaYW4BqAGNMERDh06iUCgRuF3xwC7z/P3YY+vAftt+x4zJa6UIUCItuv/PVq+/+fHxw88P4lQoS3iQwl4iE4tQxFJE0wOPTqJQKBBs/hRWvwqBpcNmbEBrefseOS2+5CzEi1k5K2d7Sh0Bkon298B/tf3ylOpA3fyFPAe8D6SLyR+Ab4GGfRqVUIFj7oe3qu+TfEJvavseObSmBlbf/AI56IvDLbbYlNudhrY+ogpo3oxBfA34B/AnYA1xgjHnH14Ep5Vd1tbYFNvCc9m151YvLaLkF1p43MR8qJAQm3AruGtgyx3fnUcrHmp2RGUBEQoBVxphhaBko1ZlsnWcrbgzx0cTjcen2huWacog8ZLShq7L9RyAeqveJtitxw6dtn9NMqQDRYgvMqcCxUkR6dVA8SgWGtf+ByIS2zbbcFvU3Mzc1EtGXXYj1QsOh72Stj6iCWostMEc34HsRWQxU1C80xvjoq6lSfuaug/UfwYAzIMybsp9HoGE1jpRDKrPVVkBUOwzXb03342yiriyEmBTfn0+pduZNAnvI51EoFUhWvg5VhTDYh9/RWqrGUVthSz/5WtcR9nnvauhziu/Pp1Q7azWBGWPmdUQgSgUEY2D2H+yUJ/1/4LvztFTQt7bC99fAALqNtM97VmoCU0HJBzeaKBXECrfYVtHE2yHcBzcS14tNBQlpOoFVl9rrb74WmwoJmbBrie/PpZQPaAJTqqGdi+xzz/G+PU9IKMSkHt6F6PHY0Y9RHZDAAAadDes/hrJmqoIoFcC8SmAiEi0iA30djFJ+t2MhRCVC2iDfn6upahy15YDpmBYY2ClWPC5Y92HHnE+pduRNNfpzgRXAp877USKin3Z17CnaBstegqyTfVPG6VBx6YcPo68ptc8d1QJLG2Sv923WG5pV8PHmr/RBYBxQDGCMWQFk+S4kpfygrgbeu96+Pu7KjjlnXFco3dN4WbWTwDqqBSYCvSfCjgV2AItSQcSbBFZnjCnxeSRK+Yu7Dl48C3IXwxkPw8CzOua8iZlQvtdWva/X0S0wgF7j7W0D+3M67pxKtQNvEtgaEbkcCBWR/iLyNPCtj+NSquMsfdHOuHze0zDhlo47b2ImGA+UNWiFHWiBJXZcHL0m2OeteseMCi7eJLCfA0OBGuB1oAS4w5dBKdVhyvbBJ7+A7EkwqoO6DuslZtrn4p0Hl/mjBdaln63K8dWj4KrquPMqdZS8SWADjTH3GWPGOo/7jTHVPo9MKV+rq4V/n21bQZPu7ZiBGw3VV9to1AJzeus76hoY2OtgU/9gh/Qve7njzqvUUfLmL/ZxEVkvIr8XkaE+j0ipjjLvESjYBOc+aVtgHa2+GkdF/sFl/miBgR3I0XM8fHaf7U5VKgh4Mx/YqcBkIB94TkRWi8j9vg5MKZ/66jH4+jEYdQWMvsY/MUQnQ0hY43vBqktBQn07H1hTROBHL9nEOevBjj23UkfIqz4TY8xeY8xTwI3Ye8J+69OolPKV0j3w2o9g9u9hwJlw7lP+iyUkBGLTGt8LVl+FQ6Tj44nvCiffY6dYWfJCx59fqTby5kbmwSLyoIisAZ7BjkDM9HlkSrW3ulp4+yrI+QxOuhN++AKEejMhgw/Fph3eAuvI61+HGnMtJPaE/94J+9b6Lw6lvOBNC+xFoAiYaow5xRjzT2NMM3OhKxXAvn4Mcr+DS16C0x/0/aSR3kjoASW7Dr7vyDqITQmPguu+hLAomP+E/+JQygveXAMbb4x50hizuyMCUsonlv7bDtoY9kMYeoG/ozkouTcUbz9YBaO6pGPvAWtKfIa9LrhmBlTs928sSrWg2QQmIm87z6tFZFWDx2oRWdVxISp1lIp3wBcPQOZYOM+P17yaktTbFvCtKrLvK/IhLs2/MYFNYB6XHZWoVIBq6QLA7c7ztI4IRCmfmf1HqCmDC58NjG7DhpJ62eeibRCTAuX5B4fX+1P6YJh4h+1G7DaiYyuUKOWlZltgxpj6uytvNsZsb/gAbm7twCLSU0TmiMg6EfleRG5vbR+l2t3a/8CqNyFrInTp6+9oDpfc2z4Xb7dVMGpKbJX6QHDy3fbesC8eaHydTqkA4c0gjqbmVfem2mkdcLcxZjAwHrhFRIa0JTiljkpdDXz+G4jLgPP/4e9ompZUn8B2HByNGCgJLCoBLvyn7Upc9Za/o1HqMC1dA7tJRFYDAw+5BrYVaPUamDFmjzFmmfO6DFgH9GivwJVq1ZIXbMvmwumQ1NPf0TQtKsHe0Fy0/WBFjrgM/8bUUEof2wpb9ZZOt6ICTkstsNeBc4EPnef6x2hjTJuqnopIFnAcsOiIolTqSKz9D3QbBX1P83ckLUvqZRNt+T77PlBaYPVGXgr56+Hl86Fwq7+jUeqAlq6BlRhjthljLnOue1UBBogTkV7enkBE4oD3gDuMMaVNrL9BRJaIyJL8/PzDD6DUkVg3007S2PdUf0fSuqTetgVWn8ACYRBHQ0MvtDFtnQfvXGNvCFcqAHhTieNcEckBtgLzgG3AJ94cXETCscnrNWPMjKa2McY8Z4wZY4wZk5YWAMOHVfBbMwPeuspWtBh7vb+jaV1ybyjZefAaWGyA/R1EJ8HPl8B5z8CeFfCHNPj4Xu1SVH7nTR2dP2AHYcwyxhwnIqcCl7W2k4gI8H/AOmPM40cXplJeqCyEVy+G3csgfShcPRNiu/g7qtalDoS6atj4KUSnQFiEvyM6XFQiHH+Vnbl50yxY/JytJdlvir8jU52YN6MQXcaYAiBEREKMMXOAUV7sNxG4CjhNRFY4j7OPJlilmlW2z3Zv7V4G/afC9bODI3kBDDnPVp/fvRzSBvk7mpZNvB2ueNd2KX75O6gp93dEqhPzJoEVO9exvgJeE5EnsUPkW2SM+cYYI8aYEcaYUc7j46MNWKkmfXQX7FwE5zwOV7xja/oFi6hEOPPP0GMMnBsE9QfDIuG0+2x34qwH/B2N6sS8SWDnYwdw3Al8CmzGjkZUKjBsXwDr/wuT7oGxP/N3NEdm9NVw/ZeQNtDfkXhn9DUw7n/srQrbF/g7GtVJeVPMt8IY4zbG1BljXjLGPOV0KSrlX8bYbrf/3gHx3WD8sVXuaP6m/SzcUkBFTasdHv5x2n2QnA1v/8SWwFKqgzU7iENEyrDD5g8sct4LYIwxfpzzQXVK1SV2OpSvHrPJS0Jgx7e2evuPXoKIDp7F2Eeqat3c8+5KPlplq7nFR4Xxn1sm0ictzs+RHSIqEX78Kjx3Cnz6Szu/mlIdqNkEZoyJ78hAlGpRXQ28ciHsWmrfRybYubNGXApnPWKHegeRnYWV5OSV0TM5hqjwUFLjIvli3T6WbS/i8+/3srukmon9unDp2F784t1VPDN7E4//2JuxUx0sYwhM+gXM+QMMmgbDLvJ3RKoT8Wo6WhE5CehvjHlRRFKBeGOM3pKvOs6nv7LJa+Id0PMEyBxjuw/7/QBCvLmUGxi2F1Tw4vxtvL54B7V1nsPWhwiM6pnEo5eMZGK/VABW7izmhflbuW1Kf7JSA6yaPsDJd8Hyl2H5q5rAVIdqNYGJyAPAGGAgdnbmCOBV7DB5pXyrthK++K0dLDDxdvjBQwfXDTjDf3G1UU2dmw9X7Ob+D9bg9hhOH5zB6N7JbMorp1eXGIoqajm+dzKnD84gIqxxQr5hUh9eWbidv8/ZxKOXjPTTT9CCkFAYdSXMfRieHAlXzgjMyv/qmONNC+xCbB3D+sK8u0VEuxeV79VW2huTd3wLY6+D037r74iOyBuLd/DnT9ZTUuVicLcEXrxmLF0TvR/mn54QxaVje/Laoh388qxBpMZF+jDaI3Ty3bZe4vcz4LVL4LpZdn4zpXzIm76XWmOMwRnQISIB2IehjjnGwJw/2uR17lNwzl8h1Kse74Dhcnv48yfr+fWM1QzsGs9L145j5q0T25S86l0xvjd1HsMHywN0Xq7QMLjkRbj2MyjJhX+dBvvW+jsqdYzzJoG9LSLPAkkicj0wC3jet2GpTq1wK3xwEyx4xt5vNPpqf0d0RJ77agvT520mIyGSZ68czSkD0ggLPbLrdQMy4hnVM4lXF25v8tpZwOg13iayoq3wn5vBE8CxqqDnzX1gjwHvYovyDgR+a4x5yteBqU6qtgJeOBNWvgF9ToVz/ubviI7I0u1FPPllDhP7dWHWXaeQHHv09Q1vn9KfbQWVvDA/wMdPDToHLphuB9ls0OI7yne8+jpojPnCGHOvMeYeYLaIXOHjuFRnZAzMeRjK98LpD9r7ioJohGG9kkoX97yzkoyESJ669Djio8Lb5binDkrn9MEZPPVlDntKqtrlmD4z4kcQ1xW++5dWrVc+09KMzAki8msReUZEpop1K7AF+FHHhaiCXmUhzLwd/jYc/n4C5C49fJuqYph5m+02PP5qOOnOoBwEUFFTxw+nf8vOwkr+cvFIurTzgIsHzh2C22N4clZOux633YWEwsTbYMtcWPmmv6NRx6iWroq/AhQBC4DrgHuxQ+jPN8as6IDYVDCrrYQ170LxTlj6b6jIh5Aw8Ljg+dMgfQic+r+QnAWuKvjsPshdbIfKT3nQz8EfmfV7S3nww+/ZnF/Oiz8dx4S+7V8Nv2dKDBeM6sGHK3dzy6n96JkSwBPLLK8AACAASURBVNVHTrjRTiz6yS/srNjxGf6OSB1jxDTTvBeR1caY4c7rUGA/0MsYU+arYMaMGWOWLFniq8OrjlJTBu/+DHI+AwS6j4IzH4FeJ0DeelvBfOOnh+93wXQY1epUcwHHGMMTs3J4Zs4m3B7D/ecM5rqT+/jsfNsLKpj29Df0SonhP7dMPOKBIR1i/yb4xwkw5AIYfK69dy882t9RqSAiIkuNMWOaWtdSC8xV/8IY4xaRrb5MXuoYUF1qW13z/gJle+CUX9pv4Q27AtMHweVv2W/mq96CsChY+6FNbiMv9V/sR+FfX2/hyS9zuPC4HvzyzEFHNEy+LXp3ieVPFw3n1teX8+L8bVw/yXfJ8qil9rNV6xf+3X42+pxqR5X2nghx6f6OTgW5llpgbqCi/i0QDVTiw2K+2gILUivegNXvwOYv7fv0oXDa/TDwLBBpff+6WnvNJCTUt3H6wNb9FUz92zymDMrgH1ccT0iIFz9vOzDG8D+vLGX2+jweu2Qkpw1OJzwkhOiIAPwdul2w5EXIX2crqoCdwPOCf0LfU21RYKWa0VILrNkE5g+awIJE6R6ITobKAnj5fCjIgdg0u+6Sf9tv194kriBXU+fm8n8tYv2eUubcM5n0hI6dRLO02sX5z8xn6/6KA8uuOTGL304b0mGJtE2MgR0LYP9G+Ox+qC2zieyif8Hgaf6OTgWoI+1CVMHC4/H9cPP9ObD4Odi52M7EG50C7lqoLYfjroRpT9hBGp0gcdWbsWwXS7cX8ddLRnZ48gJIiApnxk0nMnt9Huv3lrK7pJp/f7uN0moXD543lIR2Gr7fbkSg94n2Mfg8OwnpvEfhrSvg9Idg/E12tmelvKQJLJhVFcFrP7JV2qMSIC4DRl0Bwy6GxB4Ht/O4IW8tZAzzPsHU1dradl362a7B92+yCSttAIy51p4zNALOfRIyhvrm5wtgJVUuHvl0PSN7JnHR8T1a38FHkmMjuHh0JmC7FfumxfHM7BxKq+p4/uomv7QGhpgUOP4ndmDHOz+1A3v2rbGtsU70JUgdHU1gwaiuBvI32BZR7ncw9mdQsR/2rIQvfgNf/s4mnj6TIe972PqV3W/weTDpXjt31rqZthuw+3EgoXb7favhu+fBeOw08YWbD54zvjtcPdNelFc8O28zxZUuXrtuGBIg/+GKCHf9YACRYSE8+tkGvli7j/F9UoiLDAuYGA8TnQxXvQ+zfw9f/9V+2Zr6h8ZfwJRqhl4D8ydXlb3AnbcW0gZ5NyljXS28MNWW6QEY8WO46Dn72uOxy1e9Ces/htJc260XlQhRSbbIqrum6eOGx4KrAsKiISzCThg5+moo2AxZJ9lWXScd/rxxXxkLNheQk1fGlvwK9pRUs3V/BeeP6s6Tlx7n7/AOU1XrZuoT89hZaKt19EiKJiU2gtG9k9lTUsWX6/IwwHUnZfPLMwcFxvUyj9u2wr592r6fdK+9mT1Ca4d3djqIIxDlrYOXzrU3+AJExNnCtSfdBbHN3ADrccN/boWVr9tvqWmD7OSOUU0MCC3Ph+/fhxGX2G+5YIvkfv2YvZg+4Az7XLQN1n9kv/F2HQHHXQWxdiLFzt6VU1Xr5uGP1/H64h24PfbvJDREOKlfKlOHZvDD0ZlEhgXgqD9gT0kVs9buo6ymjqXbisgvr2FVbgkpsRGMyEykts7Dt5sLOCE7hZeuHUdUeID8HNvmw+s/tgM8UvrApW/YWy8AqktgzXv2M586AEID7Bqf8glNYEejeKe96TYsys42ezTfCKtL4Mvf2xbXjgV2IMSwi+0Ivvz1NuH0mgBXvAMRTVRYeO96WP22nTzwgr8feRyqVct3FHHr68vZVVzF5Sf04ubJfUmPjzpssslg4vGYA60tYwyvLtrBbz5Yw6kD03j8R6PapeBwu6guhZzPYcYNtgchfbC9ZlawGYq3H9zuxJ/bwR9BePuF8p4msKasfAs+/SVknWwHIjRVd2/Zy/DFA1BVaN93HQ4/egVSsg/ftn5oeXgTo9E8bvjoblj6on2fPhQyR9tukqReB7db9Bx8ci8k9oRbl4Cr0taSK90Fa/9jr3cdd6WdH0v/aH0iv6yG295YzoItBfRIiubxH43khD7tXxIqULy6cDsPzfye7NRYnrz0OAZ3a/fbO4/cvrUw43o7uANsN/e0x21tzWUv2S99g6bZgR9NfeFTxwRNYA2V7ILXfmhbQfV6jretnqgEO3jh68fsKLuqIlur75KXoHQ3vH8jJHSDKQ/YpOKqsolv93I7+CE5y5ZDyhxjR+ytfNPut+ptKNlhz3X1fyH75KZjM8ZeyJ79e9ulWFcNnjq7Lq4rZE20N3/qUON25/EY7vtgNW8s3klkWAiXn9CLK07oTb/0OH+H5nPzN+3ntjeWU1nr5p0bJzCsRwDdWGyM/Tvbu8p+sasf3GEMvHAG7Fxk/1aGXQxn/aXpL5AqqGkCc1XbhDP3z/beE1elLRo77n/sH8B719mJ+IZeCLMegsh46DfFXhMa+7ODrZ1NX8KbV0BdE1NZ9Bhjz1G2xxkQUYkziTUkZ9uW1sX/B3Fprcc7/0nb1RiXbq+LDbsYuvRtr9+GOkRlbR33v7+GGct3cdawrtx+en8GdQ2glkgHyCurZtpT3+Bye7hsXC9unNw38O4jO5SrCrbPt5Vg1rwLsemQNtBeu/3B7yCxV/PXk1XQ6DwJzFVlu9p2LYUt8+y8UvXf4DwuCAm3Q8t/8FDje5cWPQuf/BIwEN8Nrvuy+WG85Xmwd7UdMZjUG9Z+AFmT7P1RlYWw4RPb1bdrKZz+gL3gHBnf9p+ldI/9Nlk/AEO1yBjTaKj4ip3FzFq7jzOHdWVo9wREBJfbw8Z9ZSzeWsj2gkrW7i5FBBZtLUQE7jx9AD8/rV/gDjn3sRU7i3n443Us3mq7zLO6xJAUE8FPJvTmlAFpGGDZ9iLmbsxnYEY8l47rGTiDWHJmweJnba/J3tW29yKmC/zsC/3yF+Q6TwL76jHb/QZ2BFP2JFuqxjjTmo+7ofkPc3k+bPsKuo3SD3yQqKp18922Qj5cuZs56/PomxbHtJHd+DpnP1+s3Xdgu8TocBKjw9lTUoXLffjn/ZwR3bh6QhbjsoNv/jFfWJVbzEer9rCjsJKt+ytYv7fpGt5JMeG8cf34wLpuBrD9Wzsgas179raQgWdCTKrt0dj6lR29GB4NEfF29oPUgfbWkabUlNkvpok9g3Jy1WNB50lgFQX2xt2eJ+h1omOE22NYtLWA5TuK6RIbwZDuCazZVcr8Tfv5Kiefsuo6EqLCiI8Kp7iylopaNyLws4nZ/GhsTxZvLWTR1kJy9pUxNiuFwd0SiAoP4ezh3YgMC6Gi1k1cpN7P3xyPx/D52r3kldVQVetmaPdERvVK4u63V/DZ9/tIjA5nxs0n0jctAK8V7l5hB4GU59kRwBib0Oqq7M37xm23C4+F0+6D4ZfYAVcLnoFtX9v7JqtLbe8N2FHDWSfZm/4HnmW3jUqEjCF++xF9wu2yl1q8mTGgdLfdPrm3z8LpPAlMHVOqXW6ueXExC7cUHrYuPT6S3l1i+NlJfThlQBrREaFU1brZW1pNYnQ4KYEyJPwYtqOgkov+OZ/w0BCevuw4xmQFcAt27xrbvdhznE1MKX2gaCts/Nz22tSWH9xWQiBzrL1lJrGn7cWpLACMvUThrm287bQn7L2TcRmwa5m9BrfkBdtyi0mxJdcm3Gr3yxhqr623pTVnDNSUHnnV/vI8+OAm+/OPvMy2QHufZKvqGAML/2ETkccN276xxbnrqu2+3UbZKZE8dfbyS1LPg8dd9bb9goDYsQLHXWXn/quthB3f2p87e9KRxdyAJjAVdEqrXTz04VreW5bL784fynkju1NU6WLD3lLCQ0M4ZUBaYE/k2Ems2VXCja8uJa+0hp+dnM29UwcGRmWPtqgqhoJNttVRugdOvtte026KMTbZrXzTtupWv2OH8x8qrqvtpqwptYmhuvjgushE6HG8TQp9T4WEHrZOaddhhx+nosAWO965yCaQ/Tk2CRZtA8ROUZM9ySkF9z0kZtrZr7d/a+Ms2mbLztUnpHohYXbbou0cGGwG9raijKF2LICnzs7Zt3+jXRcaYVupfU+zo0KXvWLHCvQ8Ab77P3ucIRfYa5D1Zeh+Ngt6jvXiH6F5fklgIvICMA3IM8Y08S9zOE1gCuCD5bv448fryC+r4ZoTs3jwvM5XLDiY5JVV86eP1/P+8l2c3D+Vod0T6ZUSQ9fESOKjwsktqgRAEAyG0b1SSE+IDJzqH0ejfiRkWJRt6aT2t92UST1tkkBsy2fRdNtlmZxtk9Hu5VC8wyY4sNte9qZNIOFRsPXrg7M/VBXZyiP7VtuapNUltuxbTBfbMozpYpNwfZdoQ0m9YcCZMOan9nVJrh1V/d3ztnZqZLzTJXq2jbnhfalgk+/ORTaZrZlhE5rL/nuSNthOn5Q+CMr2wUd3webZ9jzZJ9uR2cN/eNT3rPorgU0CyoGXNYEdW2rrPGzYW8a6vaWUVLqYOjSD3l2Ovmbdp2v28uSXOazbU8rgbgn84YKhHN8rudOOCgwmxhie+nITbyzeQX55zYHSWy25dGxP/njhcEKDrcXWXtx1UJFnk9CM/7HX7xFbmaciz77OOslWG8kcba81hYbbViDYpFK0zSZNdx2Ehtn7XPPX267PuHTv6qu2RXWJTa7dR9kWXAfwWxeiiGQB/9UEduxYsLmAe99dSW7RwXvhEqLC+OCWifTx4kL+ml0lzFy1m417yyipcjEmK4Wo8FDmbshjVW4JAD8e05P7pg0O/PuQVJPKql3sLq5mc345LreHrC6xVNTU0SUuEo8xTJ+3mZIqF3M35DOoazw3n9qPc4Z367yJDOw1qG/+ZpNT6W7oNgJOvE0rjBDgCUxEbgBuAOjVq9fo7du3N7ep8rNnZufw2OcbyeoSw50/GMCIzCRcbg+XTF9AlcvNoz8cwfmjGt8/V1Xr5tPv9zBrXR7Ltxexu8T2xYeGCGlxkewrq8YYyEiIZMrgDB44d0jg3FukfMYYwxuLd/Lwx+sor6mjS2wEBjs4p1dKDOeN6s60Ed0PbF/tcrNuTynbCypxuT30SYvluJ7JwXe9TbVZQCewhoKpBeZye9hbUk3PlGPvG1JZtYu3l+SyZFshMRFh9M+I49vNBXy1MZ8LRnXn4YuGExNxcOj5prwyfj1jNd9tK+Ki43qws6iScGeAxbebCwCICg/hpH5pjMtO5sdje5EYbVtXbo+hps7d6Hiq86hze/hi7T4+X7uPqPBQNu4rY+n2IgB+flo/kmMiiIkI5YX5W9m4r7zRvsN6JPC/Zw3mxH6p/ghddRBNYO3EGMPMVXv4aNVuvt1UQFlNHZnJ0ZzUL5Wl24sIDw0hOy2WHwzO4Ozh3XxSuXxnYSUb9pbRLz2OXikxXn0DrXa5+ez7vSzZVkR2aizTRnQjPeFgzbg1u0q4460V5JfVEBcZRn5ZDbVuT6NjRISFcMfp/blxUt8mz1lT5+Y3H6zh7SW5AMRHhpGVGktqXATTRnTn3JHdg7qSu+o4NXVubn51GV+uzzuwLCo8hIfOG8rQ7onER4WxaGshT87KYVdxFWN6JzM2O4WRmUn0SYulb1rcUXVHuj2GjfvKiIsMIzM5Wq/B+lmnSWCrcot5Z0kuo3omMTYrhcjwEFJiIw60BlqzZlcJ1S43I3smkV9WQ1iIkBwbwcZ9ZdTWeXhm9ia+XJ9HUkw4J/VLJTM5hi/X7WNvSTUjeyZR5XKzo7CS/LIaBmbEc/fUAYzLTiExOvywP4KPVu3hgxW7SIoOp296HKcPTqdvWlyzfyxuj+Hp2Tk8MSvnwLLQEOGMoRn0S4ujrKaOrglRDOuRyIl9uxwonfTErI289O12ymvqiIkIpbLWTYjAqQPTufyEXry2aAdzN+SREhvJaYPSKCivpV96HGcP78aITHvfyb7SGqIjQg+0mppjjGHNrlK6J0XRJU5vJFdHzu0xrN1dSkZCJPtKa0iLj6RrYuNCvdUuN0/PzuHZeVuoazBopEtsBNmpsZzUP5XTBqUzqGuC11+eFm8t5J53VrKj0I60G5uVTFJMBGnxkaTERHDVhN5kJDRdMNgYQ2lVHbnFlSREhdMjKZo6jzlwbmMM5TV1Tc6QbYyhzmO8/r+qPR1ahq2ipo7CiloSY8IPXIeuc3sQkUZfDEqrXTw3bwuVtW66J0UxoW8XhnZv/0LQ/hqF+AYwGUgF9gEPGGP+r6V9jjaBvbHYzm/U8MMcIjCmdwrbCioYk5VMt8RoiiprCQsRdhdX4zGGzfnlVNS4Ka+xld8jw0KoqbMtkMTocEqq7J34EaEh/O/Zg7j6xKxmE43HY/hgxS4e/ng9+8vt7MfdEqMYmZnE8b2TiAgN4aPVe/huWxHdEqOocrkprnQd2D8+MozI8BDiIsMwwCgnmS7YUoAxcPrgdG48pS9Ltxexo7CSD1fspqK2DhE5MPJrRGYi2amxbNhbxvq9ZUwb0Y3LxvViQp8ubC2oYMayXF5ZsJ1Sp4rFVRN6c+X43nRL7JwzLqvgtyq3mMKKWtbvLWP1rhJW7ChmV7EdaJQcE85VE7IY1j2Bwd0S6JoYdViiqKyt4+uc/fzyvVVU1rq5fUp/AF5buJ0qlxuX21DlcpMYHc7vzx/GmcO6EhpivyS+sySXhVsK2LivrFHZrejwUKpcbgZ1jSctPpIVO4opq6mjR1I0oSHCyJ5JxEWG8sXafewvtzdHp8ZFMGlAGgMz4umeFE1eWQ0jMhPpkRRNt8Sow/7fcbk9hIeGUFRRS3xU2IF7Iz0eQ3Wdm237KymsqKWgooa5G/KJiQjFAAXlNeTklVNS6aK4ysWAjHiGdEugtNrFoi0FlFbb/wuzU2NJiA5n3Z5S6s88urf9f/S9ZbmH/TtcOb4XFx+fyYCMeGLbqcJNp7qReUdBJR+v2XPgl7dhbylfrN3HgIx4Vu8qobLWTa2TnPqmxZIcE0FSTASZydFkJtsP1trdpWSnxZJXWkN+WQ2TBqSSGB1O/4x4r0vmlFa7mLM+jw17y1i7p5S5G/IPrMtOjeXHY3vy04lZRIaFsq+0mhnLdlFQXoPbGEqqXFS73FTUuFm2w3b7DeuRyMS+qZwxNKPRDbwut4faOg8ut4dql4ePVu/hszV7Wb+3lOTYCH555iDOHt7tsPhyiyqZsWwX54zoFphlgJQ6Ch6P4ZM1e9lfXsPHq/ewaOvBai4hAmcN60bPlBhq6zzsKaniu22F7C+vpWtCFG/eMJ6s1MNvC9mUV86try9j/d4yhnRLQATW7SnFY7AtxIQoTu6fSnZqLOU1dWzJryA0RFi9q4ScfWVMHdqV7olRfL+7lP3lNazMLSE2IpQT+6UypFsCJVUuXl6wjebuQKifTbtbYhQeD+wuqeLbzQV0T4piZ2EV8VFhdImNYFdx0zU/6yXFhJMeH0lspN0+IyGKr3P2U1rtIjYijIyESEZkJpGREMXCLQVU1tbRLz0OjwcMhrkb8skvr+H8kd254LgenNw/jfyyGv72xUbeWrITsCOTTxuUzq/OGnxYy7mtOlUCa4nH+WTsLa2m2uX2ath3e6mqdfN1Tj49kqN90sw+1KHdAkp1VsYYtu6vYE9JNbuKqli+s5j/rtxNmdPjkhYfSb+0OG6c3JcTslNavMG62uXm5QXbeHnBdjKTo+maEMXpQzI4Z3i3Nv+91bk9GGjUGjTGUFhRS63bQ1GFi9jIUNbsKmVXcSUb95WzZlcJe0qqqXN7yEiIIjUuEhHISIgiJiKU7QWVJESHkRQdQXiYcEJ2lwPrBnWNx23MUY/yrXN7qHS5D7vNxRjD52v3sSmvnK825rOruIqXrh131F+QNYEppVQDHo/tEqx2ufV6bYBrKYHp2GWlVKcTEiLERoa123Ua5R86rlkppVRQ0gSmlFIqKGkCU0opFZQ0gSmllApKmsCUUkoFpYAaRi8i+cDRlqNPBfa3QzgdKdhi1nh9L9hiDrZ4Ifhi7qzx9jbGpDW1IqASWHsQkSXN3TMQqIItZo3X94It5mCLF4IvZo33cNqFqJRSKihpAlNKKRWUjsUE9py/AzgCwRazxut7wRZzsMULwRezxnuIY+4amFJKqc7hWGyBKaWU6gSCNoGJyD0iYkQk1XkvIvKUiGwSkVUicnyDba8WkRzncXWD5aNFZLWzz1Pig/lHROT3TjwrRORzEeke4PE+KiLrnZjeF5GkBut+7Zx7g4ic0WD5mc6yTSLyqwbLs0VkkfNzvCUiEe0dr3OeS0TkexHxiMiYQ9YFZMwt/CxNxuUPIvKCiOSJyJoGy1JE5Avn9/OFiCQ7y9v8efZBvD1FZI6IrHM+D7cHcswiEiUii0VkpRPvQ87yJj+DIhLpvN/krM9qcKwmP+c+ijtURJaLyH/9Hq8xJugeQE/gM+w9Y6nOsrOBTwABxgOLnOUpwBbnOdl5neysWwxMcPb5BDjLB7EmNHh9GzA9wOOdCoQ5rx8BHnFeDwFWApFANrAZCHUem4E+QISzzRBnn7eBS53X04GbfPR5GAwMBOYCYxosD9iYm/k5mo3LHw9gEnA8sKbBsr8Av3Je/6rB56PNn2cfxNsNON55HQ9sdD4DARmzc94453U4sMiJo8nPIHAzB///uBR4q6XPuQ8/F3cBrwP/dd77Ld5gbYH9DfgF0PAC3vnAy8ZaCCSJSDfgDOALY0yhMaYI+AI401mXYIxZYOxv9WXggvYO1BhT2uBtbIOYAzXez40xdc7bhUBmg3jfNMbUGGO2ApuAcc5jkzFmizGmFngTON9pHZ4GvOvs/5Iv4nViXmeM2dDEqoCNuRlNxtWB52/EGPMVUHjI4vOxvxdo/Ptp0+fZR/HuMcYsc16XAeuAHoEas3PecudtuPMwNP8ZbPhzvAtMcT6zzX3O252IZALnAM8771v6m/F5vEGXwETkPGCXMWblIat6ADsbvM91lrW0PLeJ5e1ORP4oIjuBK4DfBnq8DVyL/YZKK3E1tbwLUNwgGXZEvIcKtpibiyuQZBhj9oBNGEC6s7ytv2ufcrqrjsO2agI2Zqc7bgWQh02Um2n+M3ggLmd9CfYz25G/4yewjQeP876lvxmfxxuQs7mJyCygaxOr7gP+F9vNddhuTSwzR7C8zVqK1xjzH2PMfcB9IvJr4FbggUCO19nmPqAOeK1+t2bO39SXoHaN14mn1Zib2q2ZGDok5iPg7/MfDZ9/br0ORCQOeA+4wxhTKs1fKvZ7zMYYNzBK7LXm97Hd4c2d26/xisg0IM8Ys1REJrcSU0vr2i3egExgxpjTm1ouIsOxfaYrnQ9lJrBMRMZhs3jPBptnArud5ZMPWT7XWZ7ZxPbtFm8TXgc+wiawgI3XuWg9DZjidFfSQrw0s3w/tksmzPn2dcTxehNzM/wa8xFoKd5AsU9Euhlj9jjdbXnO8rZ+nn1CRMKxyes1Y8yMYIgZwBhTLCJzsdfAmvsM1sebKyJhQCK2i7ejPjcTgfNE5GwgCkjAtsj8F+/RXtDz5wPYxsFBHOfQ+ILsYnPwguxW7MXYZOd1irPuO2fb+kERZ/sgxv4NXv8ceDfA4z0TWAukHbJ8KI0vvG7BDjoIc15nc3DgwVBnn3dofHH3Zh9/HubSeBBHwMd8SPzNxuWvB5BF40Ecj9J4QMRfjvTz7INYBXtt+IlDlgdkzEAakOS8jga+xn5xbPIzCNxC40ERb7f0Offx52IyBwdx+C1ev/1htNMvcRsHE5gAf8f2Ia+m8X9k12IvFG4Cftpg+RhgjbPPMzg3drdzjO8551gFzAR6BHi8m7D90yucx/QG6+5zzr2BBiMgsaO5Njrr7muwvA925OQm50Me6aPPwYXYb3U1wD7gs0CPuYWfpcm4/PEA3gD2AC7n9/sz7DWML4Ec57n+y1WbP88+iPckbFfUqgaf37MDNWZgBLDciXcN8NuWPoPYVs87zvLFQJ/WPuc+/F1P5mAC81u8WolDKaVUUAq6UYhKKaUUaAJTSikVpDSBKaWUCkqawJRSSgUlTWBKKaWCkiYwdcwTEbfY2QC+dyp/3yUiIc66MSLyVAv7ZonI5R0X7WHnv82prv5a61sf8Tnuc34/Kxr8rlY45x4oInOd9+tE5DlnnxgReU3s7AhrROQbpwKGUh1Gh9GrY56IlBtj4pzX6diKKPONMQ94se9k4B5jzDTfRtns+ddj75PZesjy+soH7X2+A78r5/1nwD/MwRJjw40xq52yaGnGmLuc5QOBbcaYmvaOSanmaAtMdSrGmDzgBuBWsSY3mNfolAatj+UiEg/8GTjZWXan0yL7WkSWOY8TnX0nOy2Vd8XOp/aaU3kbERkrIt86rb/FIhLvFHF9VES+EzsX1f8cGquITMfeJPqhc+4HReQ5EfkceFnsfFIvOq2g5SJyqrPfNSLygYjMFJGtInKr0+pcLiILRSSlDb+ybjQoIm2MWd1g+a4Gyzdo8lIdztd3bOtDH/5+AOVNLCsCMmhcUWAmMNF5HYct7XRgvbM8BohyXvcHljivJ2OrbWdivxguwFaGiMCWyhnrbJfgHPcG4H5nWSSwBMhuIs5tHKw28yCwFIh23t8NvOi8HgTswFY/uAZb/SAeW66oBLjR2e5v2CK3Xv2ugJ86+38C3MnB0kejsDUFFwB/oEHJNH3oo6Me2gJTnVVTFbHnA4+LyG3Y/6ib6qILB/4lIquxZXKGNFi32BiTa4zxYMsYZWEn2txjjPkO7PxwznGnAj9xptJYhC131N+LuD80xlQ5r08CXnGOux47wesAZ90cY0yZMSYfm4BmOstXO3F5xRjzIrZC+jvYJL1QRCKNMSuwrcNHsbUDvxORpiqpK+UzAVmNXilfEpE+gBvbgjjwn64x5s8i8hG2ft5CEWmqAv6d2HqLI7EtreoG6xp2obmxf19C01NFIREcjwAAAWlJREFUCPBzY8xnbQy/4pBjNKdhLJ4G7z208e/eGLMbeAF4QUTWAMOApcZOxjgDmCEiHuzvbV1bjq3U0dAWmOpURCQNWzH7GWOMOWRdX2PMamPMI9guvUFAGbYrrl4itkXlAa7CVrdvyXqgu4iMdc4R70wt8RlwkzP9ByIyQERi2/jjfIWdJBURGQD0whZHbTcicmaDGLtiW4q7RGSiiCQ7yyOwLdHt7XlupVqjLTDVGUQ7XXXh2Ek6XwEeb2K7O5yBEG7slDKfYFssdSKyEvg38A/gPRG5BJhD4xbRYYwxtSLyY+BpEYkGqoDTsVOyZ2HnsxMgn4NTsXvrH8B0pzuzDrjGGFMjzU/geCSmAk+KSH1L815jzF4RmQr804k9BDvP3XvteWKlWqPD6JVSSgUl7UJUSikVlDSBKaWUCkqawJRSSgUlTWBKKaWCkiYwpZRSQUkTmFJKqaCkCUwppVRQ0gSmlFIqKP0/nPvUeyTBYzYAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 504x216 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%pylab inline\n",
    "import pandas as pd\n",
    "\n",
    "data = pd.read_csv(\"~/week5/tss_histogram.txt\", sep=\"\\t\")\n",
    "\n",
    "# Reset column headers to make them easier to deal with\n",
    "marks = [\"Oct4\",\"Sox2\",\"Klf4\",\"H3K4me2\",\"H3K27ac\"]\n",
    "\n",
    "colnames = [\"dist.from.tss\"]\n",
    "for m in marks: colnames.extend([\"%s\"%m,\"%s:+\"%m,\"%s:-\"%m])\n",
    "data.columns = colnames\n",
    "data = data[[\"dist.from.tss\"]+marks]\n",
    "\n",
    "# Peek at the dataframe\n",
    "display(data.head())\n",
    "\n",
    "fig = plt.figure()\n",
    "fig.set_size_inches((7, 3))\n",
    "ax = fig.add_subplot(111)\n",
    "ax.plot(data[\"dist.from.tss\"], data[\"Oct4\"], label=\"Oct4\");\n",
    "ax.plot(data[\"dist.from.tss\"], data[\"H3K27ac\"], label=\"H3K27ac\");\n",
    "ax.set_xlabel(\"Distance from TSS\")\n",
    "ax.set_ylabel(\"Relative read count\")\n",
    "ax.legend(loc=\"upper left\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Motif finding\n",
    "\n",
    "Now that we have an idea where these factors are binding, we'd like to know what type of sequences they bind to. Motif finding is a powerful technique to identify specific patterns of DNA sequence that are bound by a particular transcription factor. To use HOMER’s motif analysis program, run the findMotifsGenome.pl command using peak files from the experiments. In general the command has syntax:\n",
    "\n",
    "```shell\n",
    "findMotifsGenome.pl <peaks> <refgenome> <output directory> [options]\n",
    "```\n",
    "\n",
    "For example:\n",
    "\n",
    "```shell\n",
    "prefix=Oct4\n",
    "findMotifsGenome.pl \\\n",
    " ~/week5/tagdirs/${prefix}/peaks.txt \\\n",
    " /datasets/cs185-sp21-A00-public/genomes/GRCm38.fa \\\n",
    " ~/week5/motifs/${prefix} \\\n",
    " -mask -size 100\n",
    "```\n",
    "This will find motifs enriched in Oct4 peaks, and output the results to the directory `~/week5/motifs/Oct4`. The options at the end tell it to mask repeats when performing the analysis, and to look at regions of size 100 centered at each peak region.\n",
    "\n",
    "This will create, among other things, a file `~/week5/motifs/Oct4/homerResults.html` containing motifs that Homr found to be significantly enriched in our set of peaks. Navigate through Jupyter to view this html file in your browser.\n",
    "\n",
    "Run motif finding for the three transcription factors (don't do this for the histone modifications, which aren't likely to give good motifs since they do not recognize specific sequences like transcription factors do. of course if you really want to try it go ahead). This can take a while. You can speed it up with e.g. `-p 5` to use multithreading. Or use `nohup` to run it in the background and come back to it later.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q5)**</font> Which motifs did you find to be most enriched for each factor? How do these compare to published logos for these factors? It can be helpful to include a figure comparing them.\n",
    "\n",
    "See: http://hocomoco11.autosome.ru/mouse/mono?full=false for published motifs for many mouse factors, including Klf4, Pou5f1/Oct4, and Sox2).\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q5)**</font> Summarize the methods you used to perform motif finding in each transcription factor. What tool did you use? Which version? Where did you get known motifs for comparison?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Differential binding\n",
    "\n",
    "You might have noticed a strong similarity between the motifs of Sox2 and Oct4! On IGV you'll see these often bind together. In this section, we'll see if we can find a motif that is specific to Sox2 binding, rather than Sox2+Oct4 binding.\n",
    "\n",
    "First, merge the peak sets from the two factors so you just have one set of regions to analyze. You can use the Homer `mergePeaks` command for this. e.g.:\n",
    "\n",
    "```shell\n",
    "mergePeaks ~/week5/tagdirs/Oct4/peaks.txt ~/week5/tagdirs/Sox2/peaks.txt > ~/week5/overlap/oct4_sox2_peaks_merged.txt\n",
    "```\n",
    "\n",
    "Now, you can compare read counts from Oct4 vs. Sox2 in these merged regions. The `annotatePeaks.pl` tool can help us get these read counts:\\. e.g.:\n",
    "```shell\n",
    "annotatePeaks.pl \\\n",
    "   ~/week5/overlap/oct4_sox2_peaks_merged.txt \\\n",
    "  /datasets/cs185-sp21-A00-public/genomes/GRCm38.fa \\\n",
    "  -d ~/week5/tagdirs/Oct4 ~/week5/tagdirs/Sox2 > ~/week5/overlap/oct4_sox2_scatter.txt\n",
    "```\n",
    "\n",
    "Take a look at the output file (`oct4_sox2_scatter.txt`) above. The last two columns should have normalized tag counts for Oct4 and Sox2.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q6)**</font> Make a scatter plot comparing normalized read counts for Oct4 vs. Sox2 that you got from the merged peak set. Highlight peaks (using a different color) that correspond to peaks with two-fold higher binding in Sox2 vs. Oct4.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q7)**</font> Repeat motif finding, but this time using only the peaks that were unique to Sox2 (fold change >2). Report the motif you found. Is this different than the motif you found when analyzing *all* Sox2 peaks? Does this match to published motifs for Sox2?\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q6)**</font> Summarize the methods you used to analyze peaks specific to Sox2 (and not Oct4). What tools did you use? How did you extract Sox2-specific peaks?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. To bind or not to bind (challenge question)\n",
    "\n",
    "***Note: in this section we are leaving it up to you to figure out how to use these tools! You'll likely need to do some data wrangling along the way with some of the UNIX commands we used to get this to work.***\n",
    "\n",
    "Hopefully you were able to characterize the sequence patterns (motifs) that your TFs are likely to bind to. It's worth noting however that just because a TF's motif exists somewhere in the genome, does not necessarily mean it will bind! (This is for many reasons. For instance, some regions of the genome are condensed and not open to allow TF binding). In fact, we're likely to find many matches to these motifs in the genome just by chance, even if the genome was just a bunch of random nucleotides, and the vast majority of those will not actually be bound by our TF.\n",
    "\n",
    "As an example, let's look at the motif for Oct4. We'd like to explore: (1) how many times the Oct4 motif occurs on chr17 and (2) what percentage of those occurrences are actually bound? (i.e., overlap a peak in our ChIP-seq data?).\n",
    "\n",
    "To help you out with this, we have included the following files in the `week5` directory which you may find helpful:\n",
    "\n",
    "* `GRCm38.chr17.fa`: the genome sequence of only chr17 for the GRCm38 assembly. Useful since we only are analyzing chr17, so should be faster to analyze this than the whole genome fasta file.\n",
    "* `OCT4_motif.meme`: the Oct4 motif in [MEME format](http://meme-suite.org/).\n",
    "\n",
    "You may also find these tools helpful (already available on datahub):\n",
    "* `fimo` can be used to scan a genome for matches to a specific motif. The files listed above can be used as input.\n",
    "* `intersectBed` can be used to intersect sets of genomic regions.\n",
    "\n",
    "Remember, you almost always just type the name of a command to get a help message about how you should use it and what the options are.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q7)**</font> Summarize the methods you used to scan chr17 for matches to the Oct4 motif and overlap with ChIP-seq peaks for Oct4.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q8)**</font> How many instances of the Oct4 motif did you find on chr17? How many of those were actually bound (i.e. overlapped a peak in our ChIP-seq data?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Discussion\n",
    "\n",
    "**<font color=\"red\">FOR YOUR REPORT: See questions Q1-Q3 in the report</font>** Specific prompts are given in the report notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "To read more about the data used for this lab see: https://www.ncbi.nlm.nih.gov/pubmed/28111071\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (References-Q1)**</font> Provide at least 2 works that you cite in your report. For example, you will likely reference Homer and the paper by Chronis et al. where the data is taken from."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Homer:\n",
    "\n",
    "Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
