{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "f1b9f91e6aa74d42d12079a22dd97a70",
     "grade": false,
     "grade_id": "cell-127d0f0e7d7df377",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "# Week 4 Lab: You are what you eat (differential expression)\n",
    "\n",
    "Skills: differential expression, enrichment analysis\n",
    "\n",
    "## Intro\n",
    "You would like to study the effects that eating a high fat diet might have on your health. To explore this, you perform an experiment using mice. You feed three mice a standard \"chow\" diet and another three mice a high fat diet (\"HFD\"). After 7 weeks, you collect liver biopsies from each of the six mice and perform RNA-sequencing. You'd like to analyze the RNA-seq results to determine which genes changed their expression, and what those genes might be doing.\n",
    "\n",
    "In this lab, we'll go through:\n",
    "1. Alignment and visualization of RNA-seq data\n",
    "2. How to quantify gene expression based on the alignments\n",
    "3. Differential expression analysis\n",
    "4. \"GO\" Enrichment analysis to characterize differentially expressed genes.\n",
    "\n",
    "### Note on new lab notebook format\n",
    "This week, we'll be transitioning away from the 2-part assignments we had for the first three weeks. Instead, we'll us a new format with a single assignment. You will see 2 notebooks:\n",
    "* `CSE185-LAB4-INSTRUCTIONS`: (this notebook) will provide instructions (and hints) about the analyses you'll be performing that week. There are no points associated with this notebook. Instructions marked in <font color=\"red\">red</font> point out specifically what will need to be included in the report notebook. Question labels here (e.g. <font color=\"red\">Methods-Q1</font>, <font color=\"red\">Results-Q1</font> etc. should match up with question labels in the report notebook.\n",
    "* `CSE185-LAB4-REPORT`: This will be structured in a lab report format, with prompts for what needs to be included and how many points each section is worth.\n",
    "\n",
    "In general, we will not be asking you to paste specific commands you used to get your answer. However, it is a good practice to keep a document where you keep track of exactly the commands you used for each analysis, so that you can come back to it later and remember what you did. \n",
    "\n",
    "The goal of this new format is two-fold: (1) to give you more flexibility and independence in running the different tools (We will still provide examples in the instruction notebook), and (2) to get ready for the format you'll be using for final projects (and also for doing analyses and writing publications in general) This week's lab should have fewer actual things to code than previous labs.\n",
    "\n",
    "### Summary of tools covered\n",
    "In this lab we'll be using (or at least referring to) the following tools:\n",
    "\n",
    "* [STAR](https://github.com/alexdobin/STAR) for aligning RNA-seq reads to a reference transcriptome.\n",
    "* [IGV](https://software.broadinstitute.org/software/igv/) a genome browser. Used for visualizing RNA-seq alignments (and other types of genomic data)\n",
    "* [RSEM](https://deweylab.github.io/RSEM/) for quantifying gene expression. RSEM takes aligned reads as input and outputs expression levels for each gene.\n",
    "* [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) for differential expression analysis. This is an R package. It takes in gene expression levels across multiple replicates of multiple conditions and determines which genes are differentially expressed.\n",
    "* [goatools](https://github.com/tanghaibao/goatools) for performing gene ontology enrichment analyses. This is a Python package but also offers a standalone tool for running GO analysis.\n",
    "\n",
    "Note: Because DESeq2 is only available as an R package, we'll be writing a small bit of R code which will mostly be provided for you here in the instructions.\n",
    "\n",
    "And as usual, we'll do some plotting with the [matplotlib](https://matplotlib.org/) Python library (or, whatever method you want to use for plotting. You do not have to use matplotlib. If you are an R person, you can also write R code directly in Jupyter notebooks as we'll see below).\n",
    "\n",
    "### Summary of data provided\n",
    "As usual, data for this week can be found in `/datasets/cs185-sp21-A00-public/week4`. You should see:\n",
    "\n",
    "* `fastqs/`: This directory contains a fastq file for each of our datasets, `Chow_RepX.fq` and `HFD_RepX.fq` where `X` is 1, 2, or 3.\n",
    "* `genomeBams/`: This directory contains BAM files based on alignment of reads to a *reference genome* (more on this below).\n",
    "* `txBams/`: This directory contains BAM files based on alignment of reads to a *reference transcriptome* (more on this below).\n",
    "* `tdfs/`: This directory contains `tdf` files which we'll use for visualizing alignments using a genome browser (IGV. optional this week but a good idea since we'll be using this again next week.)\n",
    "* The `RSEM/` and `STAR/` directories contain indices needed by those tools, similar to the indices used by BWA MEM in lab 1.\n",
    "* `GOAnalysis/`: This directory contains reference files we'll use for our Gene Ontology analysis.\n",
    "* `GRCm38.75.gene_names` contains a mapping between ENSEMBL gene ids (`ENSMUSG...`) and gene names (e.g. Cdc45, Klf6)\n",
    "\n",
    "We have also added a mouse reference genome at: `/datasets/cs185-sp21-A00-public/genomes/genomes/GRCm38.fa` (build GRCm38)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "931efbcbe344d88749876e5092f71fb5",
     "grade": false,
     "grade_id": "cell-83768d57c0c64b08",
     "locked": true,
     "schema_version": 3,
     "solution": false,
     "task": false
    }
   },
   "source": [
    "The following code loads everything needed for plotting inline, and for loading R. To use R in Jupyter, you'll first have to install the `rpy2` Python module. You can do this from the command line using `pip`:\n",
    "\n",
    "```shell\n",
    "pip install --user rpy2\n",
    "```\n",
    "\n",
    "If the following cell gives you an error, it is probably because you haven't installed `rpy2` yet. You should only have to do this once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "bbdbb01db7c764bad24d6212db1bfe2e",
     "grade": false,
     "grade_id": "cell-8f70d7b1e7a229bf",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "# Run this to allow plotting in the notebook\n",
    "%pylab inline\n",
    "\n",
    "# Run this to allow using the %%R cell magic\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "# Suppress warnings (most notable from rpy2)\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "3786295217fe6fa03947bb9a606d62af",
     "grade": false,
     "grade_id": "cell-b4d774e3e87ff64e",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "# 1. Alignment of RNA-seq data\n",
    "\n",
    "As for most NGS analyses we'll be doing (besides assembly), the first step is to align our reads so we know where they came from. As we discussed in lab, we cannot simply align reads to the reference genome using something like BWA-MEM, since this does not take into account \"splice junctions\" that will be prevalent in our RNA-seq data.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q1)**</font> Summarize the dataset we are starting with: how many samples are we analyzing, where did they come from, how many RNA-seq reads were collected for each sample, and what is the read length?\n",
    "\n",
    "We will be using an RNA-seq aligner called STAR for this. STAR takes in fastq files and a reference *transcriptome* and outputs BAM files with aligned reads. **Because STAR can be very memory intensive (20+GB RAM), we have already run this for you. We will still walk through the steps to run it here as a reference for using this in the future (for example if you end up using RNA-seq data in your final project).**\n",
    "\n",
    "First, similar to BWA-MEM, we need to create an \"index\" that will be used during alignment to rapidly look up where each read came from. We created an index using the following command:\n",
    "\n",
    "```shell\n",
    "# Don't run this! Shown FYI\n",
    "REFFA=/datasets/cs185-sp21-A00-public/genomes/GRCm38.fa\n",
    "GTF=/datasets/cs185-sp21-A00-public/genomes/GRCm38.75.gtf\n",
    "STARDIR=/datasets/cs185-sp21-A00-public/week4/STAR\n",
    "STAR \\\n",
    "    --runMode genomeGenerate \\\n",
    "    --genomeDir ${STARDIR} \\\n",
    "    --genomeFastaFiles ${REFFA} \\\n",
    "    --sjdbGTFfile ${GTF} \\\n",
    "    --sjdbOverhang 49\n",
    "```\n",
    "\n",
    "<blockquote>UNIX TIP: Note, in the above command we are taking advantage of bash variables. We can define variables using the syntax: VAR=value (important, no spaces!) then reference variables using \\$VAR or \\${VAR}.</blockquote>\n",
    "\n",
    "Input parameters to this command were:\n",
    "* `--runMode genomeGenerate`: tells STAR we're creating an index\n",
    "* `--genomeDir ${STARDIR}`: tells STAR where to put the output files\n",
    "* `--genomeFastaFiles ${REFFA}` pointed STAR to the fasta file for our reference genome. Here we are using a mouse genome with the GRCm38 build.\n",
    "* `--sjdbGTFfile ${GTF}` provides the gene annotations (GTF format). This file is critical for telling STAR where all the exon-exon boundaries are and where genes start and end. You should take a look at this file to get an idea of what these gene annotations look like. See the [GTF format spec](https://uswest.ensembl.org/info/website/upload/gff.html) and lecture slides for more details. This file was created by ENSEMBL build 75 (http://feb2014.archive.ensembl.org/Mus_musculus/Info/Index).\n",
    "* `--sjdbOverhang` should be set to the read length -1 according to the STAR manual.\n",
    "\n",
    "After creating the index, we can align reads to the transcriptome. We used the following bash commands to run the alignment:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "54e86c705f3c0874296ea5157b8d3b1b",
     "grade": false,
     "grade_id": "cell-4fa34f0f5e8335c8",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "```shell\n",
    "# Don't run this! Shown FYI\n",
    "OUTDIR=/datasets/cs185-sp21-A00-public/week4/\n",
    "STARDIR=/datasets/cs185-sp21-A00-public/week4/STAR\n",
    "\n",
    "# STAR options recommended by ENCODE\n",
    "STAROPTS=\"--outSAMattributes NH HI AS NM MD \\\n",
    "\t--outFilterType BySJout \\\n",
    "\t--outFilterMultimapNmax 20 \\\n",
    "\t--outFilterMismatchNmax 999 \\\n",
    "\t--outFilterMismatchNoverReadLmax 0.04 \\\n",
    "\t--alignIntronMin 20 \\\n",
    "\t--alignIntronMax 1000000 \\\n",
    "\t--alignMatesGapMax 1000000 \\\n",
    "\t--alignSJoverhangMin 8 \\\n",
    "\t--alignSJDBoverhangMin 1 \\\n",
    "\t--sjdbScore 1 \\\n",
    "\t--limitBAMsortRAM 50000000000\"\n",
    "\n",
    "for f in Chow_Rep1 Chow_Rep2 Chow_Rep3 HFD_Rep1 HFD_Rep2 HFD_Rep3\n",
    "do\n",
    "    STAR \\\n",
    "\t--runThreadN 5 \\\n",
    "\t--genomeDir ${STARDIR} \\\n",
    "\t--readFilesIn ${OUTDIR}/fastqs/${f}.fq \\\n",
    "\t--outFileNamePrefix ${OUTDIR}/${f} \\\n",
    "\t--outSAMtype BAM SortedByCoordinate \\\n",
    "\t--quantMode TranscriptomeSAM ${STAROPTS}\n",
    "    # Reorganize the output files\n",
    "    mv ${OUTDIR}/${f}Aligned.toTranscriptome.out.bam ${OUTDIR}/txBams/\n",
    "    mv ${OUTDIR}/${f}Aligned.sortedByCoord.out.bam ${OUTDIR}/genomeBams/\n",
    "    samtools index ${OUTDIR}/genomeBams/${f}Aligned.sortedByCoord.out.bam\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "8c1b013a4b8075bc709074ab0a1b3c94",
     "grade": false,
     "grade_id": "cell-9efbc9deaec880dc",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "<blockquote>UNIX tip: The for loop goes through each of our samples and runs a separate command for each, similar to for loops in Python. You may find the for loop syntax useful for running additional commands below. See more about bash for loops: https://www.cyberciti.biz/faq/bash-for-loop/.</blockquote>\n",
    "\n",
    "This command uses STAR to align each of our fastq files to the reference genome and reference transcriptome. \n",
    "\n",
    "This outputs two BAM files:\n",
    "* `${f}Aligned.sortedByCoord.out.bam`: contains our reads aligned to the reference mouse *genome* (GRChm38).\n",
    "* `${f}Aligned.toTranscriptome.out.bam`: contains our reads aligned to the mouse *transcriptome* (Ensembl version 75).\n",
    "\n",
    "If you do `samtools view` on these files, you will see the first BAM files (genome) are aligned to chromosomes like we are used to (e.g. 1, 2, etc.). The second BAM files (transcriptome) are aligned to *transcripts* (named things like \"ENSMUST00000074245\"). These transcriptome BAMs are required for the tools in the next section for quantifying gene expression.\n",
    "\n",
    "A major difference between the two alignments is that alignments of RNA-seq reads to the reference genome will contain large gaps due to splice junctions. Find reads with such gaps by looking for CIGAR scores containing an \"N\" character (gap). e.g.:\n",
    "```\n",
    "samtools view /datasets/cs185-sp21-A00-public/week4/genomeBams/Chow_Rep1Aligned.sortedByCoord.out.bam | awk '($6 ~ /N/)' | head\n",
    "```\n",
    "\n",
    "You should see reads with CIGAR scores like `38M723N12M`. This means a read matched for 38bp on one exon, spanned a 723bp intron, and matched to 12bp on the next exon. You won't typically see this type of read in the transcriptome BAMs, since in those cases reads were aligned directly to transcripts with intron sequences removed.\n",
    "\n",
    "It is a good idea (although there no points associated with it here) to visualize your sequencing alignments using a genome browser like IGV to make sure everything went well with the alignment. See the optional IGV section at the end of the lab instructions below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "85669c350b70f5b7835c2a9089f8be93",
     "grade": false,
     "grade_id": "cell-7160ca04a905a604",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "# 2. Quantifying gene expression\n",
    "\n",
    "Next, we will use RSEM, a tool for quantifying gene expression from RNA-seq. It takes in reads aligned to the transcriptome, which we obtained using STAR, and outputs estimated expression levels of each gene.\n",
    "\n",
    "RSEM requires an initial step to preprocess the reference transcriptome, similar to the index step required for sequence alignment. We have run this for you using the command below:\n",
    "\n",
    "```shell\n",
    "# Don't run! Shown FYI\n",
    "REFFA=/datasets/cs185-sp21-A00-public/genomes/GRCm38.fa\n",
    "GTF=/datasets/cs185-sp21-A00-public/genomes/GRCm38.75.gtf\n",
    "RSEMOUT=/datasets/cs185-sp21-A00-public/week4/RSEM/RSEM\n",
    "rsem-prepare-reference \\\n",
    "    --gtf ${GTF} ${REFFA} ${RSEMOUT}\n",
    "```\n",
    "\n",
    "This will generate multiple index files with prefix `/datasets/cs185-sp21-A00-public/week4/RSEM/RSEM*`.\n",
    "\n",
    "Now, we are ready to run RSEM for expression quantification. The following command shows how to run RSEM on a single sample:\n",
    "\n",
    "```shell\n",
    "RSEMOUT=/datasets/cs185-sp21-A00-public/week4/RSEM/RSEM\n",
    "rsem-calculate-expression \\\n",
    "\t-p 5 \\\n",
    "\t--fragment-length-mean -1 \\\n",
    "\t--seed-length 25 \\\n",
    "\t--bam /datasets/cs185-sp21-A00-public/week4/txBams/Chow_Rep1Aligned.toTranscriptome.out.bam \\\n",
    "\t${RSEMOUT} \\\n",
    "\t~/week4/Chow_Rep1\n",
    "```\n",
    "(we also included `-p 5` to run on 5 processors to speed this up.)\n",
    "\n",
    "We need to run `rsem-calculate-expression` on each of the six samples (separately). \n",
    "\n",
    "We encourage you to try running this yourself. However, this can take a while (>20 minutes per sample). **So we strongly suggest you first use the precomputed results in the `week4/precomputed` directory which you can copy to `~/week4` so you can move on to complete the rest of the lab.** Then go back and see if you can write a for loop to compute the results yourself using `rsem`.\n",
    "\n",
    "A successful `rsem` run will create the following output files (for each condition. shown here for `Chow_Rep1`):\n",
    "* `Chow_Rep1.genes.results`: gene-level expression results\n",
    "* `Chow_Rep1.isoforms.results`: isoform (transcript)-level expression results\n",
    "\n",
    "For this week's lab, we will focus on the gene-level results (which combines expression from each isoform of each gene into one value). Take a look at the output file. You will see (among other things): gene_id (a funny looking string like ENSMUSG...), a list of transcript ids for that gene, the length of the gene, and the estimated expression level given in both TPM and FPKM. Refer to the exercises and the lecture slides for more info about the difference between these metrics.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q2)**</font> Describe how reads were aligned and how you quantified expression: what tools were used (even if we ran them for you), which transcriptome did you use for a reference, what non-default parameters did you add? You should also include the version of any tools you ran (although in some cases you may not be able to figure out the version). Most tools will have options like `--version` or `-v` that can print out the version of the tool you have installed.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q1)**</font> Summarize the expression results by reporting how many genes were expressed (TPM>0) in each sample.\n",
    "\n",
    "Since we included three replicate samples for each condition, one way we can assess the quality of our results is by comparing replicate samples. We have provided some Python code below to get you started with computing correlations between samples and how to draw a scatter plot to compare. Since our expression values have a large range, we compare log10 TPM values rather than the raw values themselves.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q2)**</font> Provide the pairwise correlation between log10 TPM values for each sample. \n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q3)**</font> Provide scatter plots comparing TPM values in Chow Rep1 vs. Rep2 and HFD Rep1 vs. REP2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "55bd57ca4083658dc10824a636e7f542",
     "grade": false,
     "grade_id": "cell-d77a3382be3adf04",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "(0.9889751249613147, 0.0)\n",
      "(0.9843194774410498, 0.0)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chow_Rep1</th>\n",
       "      <th>Chow_Rep2</th>\n",
       "      <th>Chow_Rep3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Chow_Rep1</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.984319</td>\n",
       "      <td>0.984712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chow_Rep2</th>\n",
       "      <td>0.984319</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.988975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chow_Rep3</th>\n",
       "      <td>0.984712</td>\n",
       "      <td>0.988975</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Chow_Rep1  Chow_Rep2  Chow_Rep3\n",
       "Chow_Rep1   1.000000   0.984319   0.984712\n",
       "Chow_Rep2   0.984319   1.000000   0.988975\n",
       "Chow_Rep3   0.984712   0.988975   1.000000"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEHCAYAAACgHI2PAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3de3jcZ3Xo+++aqzTS6C5Llm3JjiM5dgDFxI4DwU7SJAR27XbTcmkx7d67QOim3cBOoef0sp/27MM+z2nLzkPbUwiB0lJqaEggtDZtIJRgK4CvsZXEDpbii2RZtu4zGmlGc33PH7+Z8eg+uowuo/V5njzy3N+R4/V7f+u33vWKMQallFL5y7bcA1BKKZVbGuiVUirPaaBXSqk8p4FeKaXynAZ6pZTKc47lHkCmqqoqs3nz5uUehlJKrSpnzpzpN8ZUT/f4igr0mzdv5vTp08s9DKWUWlVEpGOmxzV1o5RSeU4DvVJK5TkN9Eoplec00CulVJ7TQK+UUnlOA71SSuU5DfRKKZXnNNArpdQy6/aFePpUJ92+UE7eXwO9Ukots5b2Pg633qClvS8n77+iVsYqpdRatLexetzPxaYzeqWUmqPFTrXUlRXygd311JUVLsr7TaSBXim1Zs03YOc61bLYNHWjlFqzUgEb4AO767N+Xa5TLYtNA71Sas2ab8BOpVpWCw30Sqk1a7UF7PnSHL1SSuU5DfRKKZXnNNArpdQiyfUK1/nSQK+UUotkpZZd6sVYpZRaJJlVPN2+EC3tfen7Un/O1aKomWigV0qpRZJZxfP0qc50jT4wr3r9xaKBXimlcmCqGv3lWmClgV4ppXJgYo3+ctbr68VYpZTKcxrolVJrwkotfVwKGuiVUmvCXEof8+2goDl6pdSaMJcGZpldLfc2Vi9raeRi0ECvlMoLmXXrUwXkuTQwyzwozLeV8UqigV4ptaLMFrCns5gBOfOgsNp6z09FA71SakVZaZuB5EMrYw30SqkVZa1sBrKUtOpGKbWiLMZG2flWNbNQGuiVUnml2xfiiRcu8uzprkmllGv1AKCBXim1qswWrFva+2jrGSGBoanGO+mxldhGONdynqMXkatAAIgDMWPMrlx/plIqf81W4763sZpTVwfpGAjS1hNgZ315+rX5UEEzH0t1MfZBY0z/En2WUiqPNdV4qS0dpKnGO2WFTl1ZIY8/sm1cL/iUtXrBVqtulFKrSltPgJv+MG09gWln6JkBfb51+flkKQK9AX4gIgb4kjHmqcwHReQx4DGA+vq1d6RVSs1NZnDPZobe0t7Hs6e7OHV1kMcf2bYmg/1SXIy9zxjzVuDdwO+IyL7MB40xTxljdhljdlVXr628mVJq7uZafrm3sZqGKg8dA8E1dxE2JeeB3hjTnfzZCzwH3JPrz1RK5Z/5lkamcvbvvXvjmrsIm5LTQC8iRSLiTf0ZeCfwWi4/UymVPzKD+0JKIxdjEdZqluscfQ3wnIikPusbxpjnc/yZSqk8MbGUMvMn6IXWbOU00BtjLgPNufwMpVRu5SKYTveeE++f7cJrPrQQXgpaXqmUmlEugul07znx/tmqatbqAqi50kCvlJpRLoLpdO85189aqwug5kqMMcs9hrRdu3aZ06dPL/cwlMormsfOfyJyZqb2MtrUTKk8t1yNvKYrh+z2hfjS0Td46uilNddFcrlo6kapPLdceeyZ8vDPnO5CRCj1OPNi8+2VTgO9UnluufLYmQeYzPTR3sZqfMEIguTN5tsrnQZ6pVROpMojW9r78AejHGu3Gth+YHc9H7v/9vTztHIm9zRHr5TKmdRs3WDY11iFLxjhbOfQuNz9dKtWz3YO8elnznG2c2g5hp5XdEav1Cqw2ipnUuNtqvFyoJlxKZr23hFu+sPA1Kma1GuPXuzjJ5ess4DMzUPU3GmgV2oVWG157NR4DzTfGm8qNdNU4x3XSz4lFeBTaZ7mjSV43DUc3NOw5OPPNxrolVoFVnIee6qzjanGm3lReKoZeurgsLexkgPN61fN2ctqoIFeqVVgJa8AnXi2kZm2mUu6aWJfG7V4NNArtYYtRu4/cw9XuBX4a0sHZ8zFT7SSD2arnQZ6pdawheT+M3PqqT1cd9aXz5qLV0tPA71Sa9hCcv9T5dRh+lz8aqscyica6JVawxaSLplrTv1IazfPnOnCH4zy2P1b5/WZan50wZRSa8h8912dyly35zMYjDEYVk7H3LVCZ/RKrSELzckfbr2OIOxvrptz+uVA8wbKPC7N2S8DDfRKrSEL2Xd1YtfJuR4otKpm+WigV2qNmC6gp2b5/mA03TY49fhMXSfV6qGBXqk1Yrq0TSpo+4KRSY8fbr3OM6e78AUjfOz+28d1nVSrhwZ6pfLUxBn8dKWUqZRKty80KYcuCCKCIEs6drW4Zqy6EZE3i8hxEbkmIk+JSHnGYydzPzyl1HxN3EJwtiqZqR7f31zH++7eiMHotn+r2GzllV8E/hR4M9AGvCQiqQJYZw7HpZRaoL2N1eMWMk00W6ll6ozAYGhpH1jyPWfV4pkt0BcbY543xviMMZ8Dfhd4XkTuBS2GVWqpzaUOfrYZ/GybhqceF2TGA4Za+WbL0YuIlBpj/ADGmBdF5FeBbwMVOR+dUmqcxexLP13OPjWTr/C4qC11s3tLhW78scrNFuj/DNgOHE/dYYx5RUQeAv5HLgemlJpsug23pyqHnG1BU+oCbWoR1O4tFbT1BLg2EOT7F3poXFfM8Fgs3axMrV4zBnpjzDcyb4tIiXW36QQ+msuBKaUmy1x09PSpzkmz+5b2Pp493cWpq4M8/si29H2pTpITf/qCkfQiqLbeADf9YUoK7Bhj2FLlob6ySFM2eSCr8koR2QX8HeC1booP+C1jzJlcDk4pNbVuXwhfMMK+xqpxgXhvYzWnrg7SMRDkSGs3bb0BOgaCNFR6uOkPU1s6SEd/kAQGmwj3bK6gqaaY26qKeWhHzbiDwExnBdqJcnXJto7+q8DHjTEtACLyDqzA/5ZcDUwpNbVuX4gnXrhIx0CQ9969cVygrSsr5PFHttHS3ocvGKHt5ghOh/Dojlqu9I/SNRRkeCyKPxRl+/oSLtzwc6V/lOZNZeysL0+naGZL1ay2PWzXumy7VwZSQR7AGPMSEMjNkJRSM2lp76Oj35qlN9V4J1XhpNI7B5o3sLG8kB7/GM+dvY7B8NNLA/SPRIgnDIUuG1f6RglGEnNeEDVb6aZaWbKd0Z8UkS8B38Qqq/wA8GMReSuAMeblHI1PKTVB5gXZiS0KMtWVFbKlysOx9j5e6fJx16Yy3nVnLRduDLN9vZeHd9Ry8spAuhvlXGiDstUl20B/V/Lnn0y4/+1Ygf8XFm1ESqkZZQbZ4VCM4bEYw6HYlM8tLXRRVexmx/oS9jfX0dLex7kuP/WVReNSNSq/ZRXojTEPLuRDRMQOnAauG2P2L+S9lFK3lBQ6KClwUFI4/p9yqnd8IBTjg/fUp/vHL2TrQLV6ZVt1UwP8P0CdMebdIrIDeJsx5m+z/JxPAq8DJfMbplJqKpmbeZztHOKvftiGiLB9vZfvn+9BRPjI3i3pC7YzpVy0kiZ/ZXsx9u+B7wOpRF4b8KlsXigiG4FfBL4y18EptZal2h2c7Ryasu3BxMB86EQHL73RT8sb/Vy4EaCptph37qjJevY+W0sEtXplm6OvMsZ8S0T+AMAYExOReJav/Tzw+1g1+JOIyGPAYwD19XpxR6mUVOCtLR3kpj8MjC9lTD3uC0Yo87h4dEct3UMhBkbDbCov5HJ/kJJCx7Sz9GzbGKvVL9sZ/aiIVJJsZJZsauaf7UUish/onWlhlTHmKWPMLmPMrupq/R9MqbOdQ3z8H09z/vowexsrObingX2NVfiCkXGz+6YaL7WlboZDMQ633mAwGOGXd26g2lvIxnIPB5rXEwjF+PKxyxxuvT7pc+baxlitXtnO6B8H/gXYKiI/AaqB92XxuvuAXxKR/wAUACUi8o/GmA/Na7RKrWLZ5sAPnejgxYt9eAuc/N47m9hZX05bT4DDrTdo7x1Jz+4BOgaCDAQilBTYaarxUlNSAJD+jKeOXpp24xCdwa8d2VbdvCwi9wPbAAEuGmOiWbzuD4A/ABCRB4BPa5BXa9VMNe+ZDu5pIBiOcVtV8bhg7AtGGA7FqPUW4AtGuGdLJQ2VHk5fHcJhk3Tzscz0zv7muvQ+sBNpLfzakfVWgsaYGHAeQEQeEZHfN8Y8krORKZVHun0hWq/5iCXMpNl1ty/EkdZuDIYDzRvYWV/OH++/kyOt3Rxuvc49Wypp6wkgCK1dfmpL3bS8McIPLvTw2/u20rTOi8FoMFfTmjHQi8gvAE9iVdt8F6vE8h+wZvX/ay4fZIz5MfDj+QxSqdWupb2PvkCEXZvL2d9cl07jNNV4OXSiIz0rT5VKPvHCxfR9rV0+2m6O8OidNextrKTbN0Z/YAxfKMr3L9zkc++7a/YBqDVtthn9/8aqiPkZ8G6svvT/wxjzl7kemFKrXWZOPjMFU1dWmG4xnOomuWO9ly1VRfiDUb5xvIPjlwYocjt4ePs6QGjvGaW00EWpx8mzZ66TMFBV7ObRHbXL+yXVqjBboDfJmTjAd0WkT4O8UtmZ2OExM4WSyrl3+8aoLnHx0b1baesJ8OyZLoZDUYaCEQDqK4toqvHSExijtNDJ0bZe3r61kr6RMfoCEQaTz1NqJrMF+jIR+ZWM25J52xjzndwMS6nVKzMtc6B56qqWurJCyjwuvnnyGqPhGM0bB7hnSyUJY/CHorgddt6ysZS9jdW0tPdx0x/mW2eu8UbvCPdtraJ5UxmCaMWMyspsgf4ocGCa2wbQQK/UBKmZ/IHmWwucMoN/alOPvY3V/PPZ65ztGuIH53s4fnmAa4NBBOHBbdVsq7U6hqSCeYXHxfcv3KTG66alfYADzeu15l1lZbatBP9LNm8iIv/JGPO1xRmSUotrqXq4TDWTP9s5xKETHdR4C5IVM1ZOPrXVX1mRk2jMcLEngNtuYyQSQxCuDgbpCUQwmPQF2rqyQh65s5ZuX4j65PdRKhvZroydzScX6X2UWnRL1cMl9TltPYH0CtMvt1ziSGs3P3y9h32NVRzc00B1iYvTVwc53Hqd+xvXUeN1s+/2Kh7YVs2GskJ2birlt/dt5UDzegSZNHZdwarmKus6+lnMbXsapZbQUq0Azfycbl+IQ8evcvFGAJtNGInEKPU4rZWrBmJxaL3mBwyBcJyekTAlBU5u+Me4o7aER+60qmm6faFpFzwpla3FCvRmkd5HrUCrvX1tLhcNTdUY7NDxqxxr66drcJRgNEFVsZuHtq3DH4xy6PhVzncPU17koC8QZs+WCnoCYaJxw1g0ht0mGGP49DPnOLinYdJKV6XmQ2f0alZrdSPo6Q5wmfe3tPfx7JmudM69pb2Pfzp5Dd9YjBK3HW+Bg9FwjOfOXqfI7WD7+hJEBK/bSVmRk4d21PDBextoae+jcyBI/0iUrqEQnUNBYPZNupXKxmIF+p8s0vuoFWitNr+a7gCXef/exmpOXbUusB5p7QagqbaYV7r8xI0hHE0QjCQAGIvFGQ5GCUVj1qze40z3p9nbWM2RYDfv3LGOkXCcTZWFHNzTsMTfWOWrbHeYuoS1KrYFOGaMuZD5uDHmd3MwNrVCrJV+Kdn2Z7faAw/SVOOlrqwwPZP3BSO0tA/w4LYaHti2jiPnbnD+xjAuG7iddjZXeHi9Z5hQJEGBU9ixoTL93kdau3nmTBeN64oZHotxoHm9zubVosm26mYH8CWgEviciFwWkedyNyylll62/dnbegLc9Idp6wmkn7e3sZrhUAynTegasmrhR6NRElgXsGwiDAQjRKIJXDZ4221V/PEv3gnA06c68YciGGPYUmX1kV9rZ08qt7JN3cSBaPJnAugBenM1KKWWQyq4NtV4efpUJ001Xk5dGUx3lUwF/AqPi1A0RoXHBVhnAqkmZL5ghJ9e6qe8yIXHZcNpg4SBQDjGSDhGwsDWdcV89j1vHtfzZl9jFR/dd9uqveCtVrZsA/0w8CrwBPBlY8xA7oak1MItpFLo1JVBjrX3U1s6SOs1P8bcWrTU0t7H0bZe3ugd4YtH3+BP/+U85UVObg6HwRgSiQSRuCEUjhGMgMdlZyyaoLakgEqvi47BEA9vr0mPaWKzM6VyIdtA/+vAO4CPAx8RkZ9i5er/PWcjU2oBJl4wzSbop16zt7GSfY1V+EMR3rmjhpJCBxUeF//tmy8TjRt2biqjtqSA1mt+4gau+8fS72EXsAk4HTb8oRjGGBLAbdXFfOLhRg6d6ODhHTXp5/cMj3Hq6q18v1K5kO0OU/8M/LOI3IHVrvhTWBt+6/+ZakXKnClnWx6ambo5dKKDjoEg7717Ix/YXc/H//EM568P01DpoXNglCv9o8SnWD3ictgocjl484YSEPAHo1zuHwUxnLoymM7tpy60HjrRwQsXegAtpVS5k23VzbeBu4A3gJeA3wRO5HBcKsNqX7C0mLL9XWRWCk1VPdPtC3G49TrdvjE6B0apryhiQ3kh+5vraGnvo6M/SLXXzYXrw3z84mkKnXbcDhu+YIS+QJjIVFEeGIsmiMQivHJ9mAfvqOYTDzWlDxpWrn/8hdZUCaWWUqpcyjZ18/8CLxtj4rkcjJraWl2wNJX5/C6mKg9tae/jmdNd9ATGGIvEKXAOUVtaSKnHSVONl2qvm6HRMC9e7CUSS9BQUUiB087ASBiDtULQAHdvKuNyf4ChkPVPw2GD0kIXlcUu2ntGOHVlkKZ1XhrXFY+7oJt5wNIdolSuZRvozwG/IyL7krePAk9ms0G4Wri1umBpKgv9XWR2mHzfro2094zwcucgTpsdxFDhcXHqyiAvtfcTCMfSvT26fCFsCAYwBjwuqzL5ui9EwloPRanbzu21Xu5cX8q5az4cdsEfitDaNZxuKZz6/FTNPejBW+VetoH+i4AT+ELy9m8k7/tILgalxlvNC5YWO+200N9FS3sfz57uoqHKk17odLpjkKuDowjCc2e7CEXi44I8QDwOYjeUFDhIJKDM46DLN8ZoJIw92QDE6bBR6HSwobyAYLSYjoEgpYWucema1BnJvsYqrZdXSybbQL/bGNOccftHItKaiwGp/LLS0k7plgUDwfTM3lvgtBY72W1c6hvljd4RwKqgSaXiDWC32zBAlddFly+IMVDkslHucWK32fjgPfW0941wz5ZKDjRvmPIAp+WUajlkuzI2LiJbUzdE5DasxVNKzWhvY/WSzly7fSGePtVJty805e26MquHTHWxi2sDQb571rog63Xb8YciXOwZIW6swJ4wt/6BJACP004wEqPbHyIcs56zrdbLvVur+Pyv7aSsyJWuqpluVW1qFW1Le196TAv9jkrNJtsZ/WeAF0XkMtZ1qAYgq92n1Nq21Gmnw63XeeZ0F75ghI/df/uU9fS+YIS2nhFe7vQxOBImkoDhkBXYMxnG99/uH40iQCRucNrA7bSqcE5fHUxfbE19zkwWepaz0s6S1MqXbR39v4tII7ANK9D/3BgTzunIlJoHQRARBKHbF8IXjLCvsepWS+HTXVR73TRvLOXwKzdINpYkmrDSMMFIYtrNFWxAaaGDoVAMl8NOZZGLYDhBIvm52R7UFnpBWS/Oq7maS5viu4HNydc0iwjGmH/IyaiUmkI2F3Z3b6mgrTfA7i0VHGnt5psnr7FjvZfdWyq40D1Mty/IDX+IgdHxtfBOm1UDPzHIp8ooARoqC9lcVcyZq0N4XHY8Lgd3bvBw18Zy9jfXZf09FnqWs5ovzqvlke2Cqa8DW7HKLFO5eQNooFcLMpeqnGxSFievDHCu02e1+w1F6QuM8dLIGD+/OUzX0BjRZHBPBW+HgNMphCNWq4KJCpyC024jkYAtVcUEwjEaa4uJxgxNtcU8/sg2vaiqVrxsZ/S7gB3GGN0yUC2queSbs0lZDIdiDI/FaO8Z4dXrPuIJQxy4NhgiOiGS2wUKXXbGovEpgzxASYGDAoedKq+bjeWFbCz3WGcNPQGtnFGrRraB/jWgFriRw7GoNWgu+ebMlMXZziEOnejg4J4GakoK0mcFAsQTCVre6GNgOEIMayHTWHJRd2YqJm5gNBKfdBE2pcABfYEohij9oxH8oRgf3VfKzvpy7UuTJW3fsTLMGOhF5DDWvwsvcEFETgLpi7DGmF/K7fBUvptPvrnbF+Kz37tAe49V717jdfPNk9f47stdIDAWjeM2Nkwyqo+Eb83YJ8b06YI8gDFCRZGDaNxwZ10J92+rTveq18CVHa0QWhlmm9F/bklGodaMuczwMtsVZKZKDrde56YvREOFVRP/58//nKFglNMdQyQApwiBWCwdxLNd8CFAZZETj8tGTyCCx2Xn13bXs6nSk/7s1EYhoIErG1ohtDLMFuivAzXGmHGbfyd73lzP2ahU3pgY2Ocyw0s9t7bU2nz71NVBHn9kG4LgcTs50LyBmpICrg+OYiCdg49PWyA5s7pSN9tqS+gfjfCmjWXpapqpVrbqzD47WiG0Msy2MvbzQGCK+4PJx1YMXS24MqVq15944SLdvtCklbIz/b2lnntwTwMNVZ5024LdWypo3lRKaaGTj339lLW70zRkDmMdjcQZGI2wsbxgyiAPtwJXW09g3P6ySq1ks83oNxtjXpl4pzHmtIhszsmI5klzgSvT3sZqjl7s5fTVIY60dvPY/VvH/f1M/HubeAaQem5/IMznfvBzvnXyGiPhKG/0jfLcy9fT3SSn4rRBPDE5Lz+Ryw7lHjcFThsOu7Clsphj7f2UepzT/r+kKQm1mswW6AtmeGzW81URKQCOAe7kZz1rjPmT7IeXPf2Ht3JMDNbNm8q4cGOYc11DdPtC0zb5gukP2M+d66KtZ3ROSZmJ5ZQpdqy8vdsOZR4XFUVu3rNzA229AToGgpQUOmbtz6MpCbWazBboT4nIR40xX868U0Q+DJzJ4v3DwC8YY0ZExAm8JCL/Zow5Ps/xTkv/4S3cXEvhpnv+xGB9oHkD7b0jdPQHOXT8Kj2BMAf3NLCzvnzS39t0B2yP04HI9LP3bDlt1p6u8TiUFDh58jd2pS/0psaueXeVb2YL9J8CnhORg9wK7LsAF/Ce2d48ucBqJHnTmfwvJ4uuVnq9bmrrOkGmzP2uBHNNf033/KYaL7Wl1obXYB2EU73fn3/tJscvDxAMx/nCh+6e9J6ZgT+z6uZiT2BckHfZoMrrptuffcslO1Ba6MQYIRaKYLfbOHVlkFKPc9JnK5VPZgz0xpge4O0i8iDwpuTd3zPG/CjbDxARO9ZB4nbgb4wxOdlrdqXn6FNb14nIjLnf5TTX9Nd0z2/rCUzaBDsVRDsHRnmly0coGpuUxoHxB+zDrdf5+k+vEo4lGInEcdmtGX0kAQ67jZGxWFbj9LqER++sYzAY4cKNYUoLnex/y3o2lBdiMCv6/xulFkO23StfBF6czwck95m9S0TKsM4O3mSMeS31uIg8BjwGUF8//39oKz1Hv7exGl8wgiArdowzzWinOmOa7vnT/V10+0KUFDrZs6WCvkCEQ8c76AmMpdM41kKo87zS5ef5124QjiboG4kQiZtJp4HB6RLwE9gEXE4HTbVeDIZXr/sJxxLs2FCSvvhb5nGt2L8TpRaDLGX7GhH5E2DUGDPlQqxdu3aZ06dPL9l4VPZSC4VSFymnSpNNdTDIvC9zG71Sj5Ojbb281D7A+tIC6soK8AejnO8eJho3FDhtxBIJHHardfB8lLpt3L6uhP7RMB/cY10rmG/6bKWnBtXaJiJnjDG7pnt8Lm2K5/Ph1UDUGOMTkULgYeDPcvmZKjcyZ+nTpcmmKpV84oWLdAwEJ70HQOdAkJ/fCHC5b4SLPSNkSs3Yo/H5BXmA4gInv/3AVgaDkXSA/tj9t8/rvVZ6alCpmWTbprgG2IB1IbU7mbvPxnrga8k8vQ34ljHmyLxGqpZVZppmutTMVKWSHf1BGqqsFgI9w2M8/9oNnn/tJpvKPRxr7yUUiSMi2MUQn+Lkci7nmx6XjQK7DafDxlAwQrc/zFd/coVvPva2uX/hCVZ6alCpmcyYuhGRu4AngVJutTzYCPiAjxtjXl7MwWjqJj9M16Pm08+c47tnrwNCTYmbsWgCh11YV+zCF4pwbWj+m5YVOoVCp4PmTaV0DY1xtX+EaAK21RTx/f/+wKJ9N6VWotlSN7O1QPh74JPGmO3GmIeT/92BVXb5d4s4TpVHjrR285WWK5y6Mpg+C/jS0TfAQF1pAUVOO0PBCMUuGztqveysr6Dbt/CdKUsKndy5vpT33b2RR7bXsKHUzW/dd9uk52m7DLXWzJa6KZqqHNIYc1xEinI0JrXMsrnwONNzDAZjDP5QhKdPddI5MMq3TncRiSUIRmLEkmn37tgYfaMRftzWv6DFFQK47XYqi108tKOGnfXlPH2qk+FwfMpmN5pvV2vNbIH+30Tke1hbBl5L3rcJ+E3g+VwOTC2fbAJhqllZqqNkZpXNcCjKpgoPL3cO8b1Xb7K+tICEsWJuZlVkNAHReVbUABS5hMoiN9E4lBTa6fGP8eVjl/novtvwB6PsbaycMqeu+Xa11sy2YOoTIvJu4JexLsYK0IW18Olfl2B8ahFlWyI4cWXrVPY2VnPq6mC6o2RTjZdDJzqo9RZw5JUb9AbChKMJEkB/IEQ4ZlhAAc0kThvc31RDXyBMQ5WHGq+bb53u4sKNYQ6d6OCmP8yB5vVTfk9dAavWmlmrbowx/wb82xKMZc1aaI12tq/PNmXR1mM19zp0ooOakoJpg2WqrcHexmqeeOEiL1zo4b6tldxZV8LN13vSuzqNRBZ/rYYAH91327g+NSWFTgRJ7+mqPeOVssy7jl5EnjLGPLaYg1mrFpozzvb12aYs0rP1fmu2Pt171pUVpuvqH91RSzASo7q4gJGxGDabMGW95AIUOoUChx1/KMbD22sm7d2aWSOfytNrLl6p2feMrZjuIeA/LP5w1qaF5Iy7faEZ89Gp56Rm3tkEvImz9Zkcae3mmTNdvHNHDRj4zstdBMLZbt6XHbtAodPOPVsqCIzFaKjy8Pgj28Z9L5jceVJz8QYJ9sMAACAASURBVEpZZpvR9wEdjK9dMMnb63I1qLVmITnjlvY+jrX3T5mPTgVCXzBCS/sAkN3Mdi5VN11DQQZHwvzba930Do8tepqmxG3jzg2l3F1fQfOmMr5/4SYH9zRM2r8VSP85s0WDzuSVmj3QXwYeMsZ0TnxARK5N8fw1b6l7osw0a83sLTPbRhoTX/fsGaui5uCehkn92lMXXtt7RhgZizISjhGJJxiLLV6Qt2HNJtaXebDb7Gyq9DAYjHDTH+bUlcF0Dv5A8/jvPlOLBu1Xo9aq2QL954FyYFKgB/588YezukwVOJa6Rnums4GpWhKkxjpT0Evl6NtujvDZ710gFjecujpI47piWtoHcNpucvrqIMFoPJ2Gd9jimEWqqnE7hANvqQOBR3fUpnvVpPiCEZ493UVDlYeDexomzd6nO/hp/bxaq2Yrr/ybGR7768UfzuoyVeBYSXnhzIPAxAuTqdy6Pxjlsfu3Tnpt0zovwUiMrqExTMJw/NIAF7sDXB4YYXSK2vfR6OKNu8jl4NE7a3nubBet13x88N6G9MEo1SytvXckXRl0M7n5SOq7zrV9slL5LtumZr8yxd1+4FVjTO/iDmn1mCpwLGWN9lxSERPHmlq9mtnpPfV+/mCUY+397Gusonmj4fjlAToHg3T7xljEUvg0m4DTJhhjcLvs/B/vuoPvX7jJixf7ON3hY1OlZ9zvNPNicWY/ndlo/bxaq7Itr/ww8DZubT7yAHAcaBKR/2mM+XoOxrbiLXfgmC0VMfFAkHmR8kDzBmtjbI+LTz9zLp2L/7uXrhIIR3nojhr2N9dxpLWbn98cIRKP5yTIu2zgdtp5U10pD9xhjQvg2mCQe2+r4M71pVMG8czffWaJpVJqsmwDfQLYnmpPnGxb/EVgD3AMWJOBfr4W66LgxFn6xPfNPBCkFjWlesN/YHc9H9hdz6efOccLF6yu048/so2/+mEb1/1hDrd289IbfQyHovQvZl4mgwAJA3ab8OAd69IppKdPdXKuy8+B5vU6A1dqEWQb6DdP6EHfCzQZYwZFJDdRYI6ePtnJF378Bh9/4Hb2Nk29A9JKsVgXBSfO0if2n5m4WUhmb/iUg3saxv1sqCyifyTC8FiUoVBu/mrtYqVrRCCegJoSN/ub69KPay5dqcWVbaBvEZEjwDPJ2+8FjiU7WPpyMrI5euKFNnoCYZ54oQ2EFV1dsZiBbOKsPbP/zMSyw9TPzINffyDMlf5RTl4e4Aev9+APRtlcVcTASDhnM3mnTagrK6AvECFhh0e2144b08SUmJZFKrUw2Qb63wF+BXgH1hn314BvG2vXkgdzNLY52VBeQE8gzIbyghkD6UoIGouZ258YwDNXtKYOAv5glFKPk6YaL0dau/GHIoAQGIvxL61d+MfitHb6SBhrNVyBUwhHc7eXcNwYitwOEgg71nv54L0NMz5fyyKVWpisNwdP5uXvwYoFJ3NRbbOQHabOdg5x6EQHB/c0zHhxLnOT63wOGt2+UHojbIOhpX2AkkI7p6/6iMWtTbdHwzGCEatdwVJsEe8QKHLbqa8o4hMPNY7by3W277LcB2elVrJF2RxcRN4P/AXwY6wZ/V+LyGeMMc8uyigXwcQGV9PJt/zvdEHwSGs3z565zvvu3piuZDnS2s3QaJiNFYXYkoeAcDSOyPg+8YtJgLduKuVS/yjb13t5cJtVzTOXgL3c1U1KrXbZpm7+CNidmsWLSDXwQ2DFBPps5VvQmJjWSNfChyIYYxgORWlp72M4FKOtx9pHtXc4nNyvFex2CMdyM7ZCp/Cph5oQEZ4508WD22qmXJyllMqtbAO9bUKqZoDZ95tdUvl+ej/d98vcJKTbF+L3vnWWc9f8bF/v5f27NmEwHG69gdMGbqeNMo+daq+bC90BEnHIVSq+wC7UlBRSVuRib2M1pR7njGdR+f73p9RyyjZYPy8i3xeR/ywi/xn4HrCidpg63HqdLx+7zOHW68s9lJxIzdxb2vvG3d/WE+CmP0xbT4CW9j5e7vARiiY4f93PuWtDBEIxStx2fnapn8BYnOpiNx0DoyTIXZDfVlPE+3fX88F76tOBO7MVw1Sbck/3/ZRSC5fVjN4Y8xkR+VXgPqy061PGmOdyOrI5Gg7FGB6LMRzKUR5iFos1I029T1ONl1NXBjEYDjRvmPbawt7GajoHRnn+tZtUelx4XHbCIWvjj5fa+3ipvZ+R8K1Vra/dCMx7bNkQoNs3xobygklpmpmqZ/Lt2olSK0nWO0wZY74NfDuHY1mQkkIHJQUOSgrnvWnWgixWCWDqfWpLB2m95scYQ5nHlV7JCrcOBhUeF1/9yWVeuz5MOJbAaRccNhslbhtxI4xG4ou9ydOMGsoLsNuFYCTB8UuDCJfGXXidKZjn27UTpVaS2XaYCjB15Z0AxhhTkpNRzcM9Wypp7x3hni2Vy/L5izUjTb2+qcZL0zprRj9Vu91nz3TROzxG52AIAxQ6hC2VRVwZCGITq6RyqYK8Dbj3tgo+9/67AHjihYucvjrIlYFR2noDPP7INurKCjWYK7VMZmtT7F2qgSxUZq46V02uZkrPLFYQy2xrkGoLMHG7vAqPi4QxOO02NpW7GRiNUeC0E4zGsYtQ7nEwFksQjEQWPJ7pOG0QS1itDHas9/K599+V/p0c3NNAMBwjFInPuu+sUir3lifPkQNLkePN9QrNqbb+A3j2TBdHL/YRisa40h9kx/oSojFDwhgCY3Fi8TjDsTgDoxEECIRjOV8AZRehwutka3Uxn3nXHeMOfG09AfoCEaq9bnbUleAPRtMXYLPdolCrb5RaPHkT6BdjRj1bkMnmYDLTe8z2/qkDyV0bS6ktdVPhcXG5f4Ril1U1E4rFKXTYGRqNEI7F6BwIMnH3vqVKyY/FDZGRCHubCiedQWX23PG47bR2DVPqcQKz9yDSdgdKLb68CfSLYbYgk83BZKb3mNiAbGJKJrUHqj8Y5UcXe/nR671UFrsocNitPjRxQ0QMr133EY0vXQ4+xeu2E4omEAzFbgfb15fw6I5anj7VOe7gNdvGIDMdKLX6RqnFlzeBfjFO+RcaZLp9IXzBCPsaqya9x8THMoO+LxjhmdNdvG/XRj52/+2c7RziyaOXGApGCccSxBJxPE47RmA02ZtmOYRjcbxuBxsrPPzPX34TO+vLJ21RmJI6KHb7QrT1BMbdNxO9YKvU4subQD/bHqjZWGiQaWnvo6V9gAPN6ycdbFra+/jhhV4aqjzA+IPKkdZuYgnD8cuDwBsIQkmhk1A0hg2rRUE8Hl+WpciFTghFrcqauxvKKfe46Qvcuug928FRUzFKLb+8CfRtNwN0DozSdjO3C4JmklkaOTGdkcpbt/eM8MQLF3n8kW3pwLe/uY5z13ycuDpIW2+ApmovI2MxxqIGj8tQVuAgmjCMLcNsPrX3iMsh/O/37wTGVwHNdnDUVIxSy29F9atZiJNX+wnHDSev9i/bGFJBr60nMGk5fypv3VhTnC45fOH8TX71iz/h/HU/zZtK8brtVHhcXBkYZWTMirA2ETxuO5FYnKUK803riih22fC4bDhtVt+a//5w07ha+GzTY3N9vlJq8eU00IvIJhF5UUReF5HzIvLJnH1W8qvIMh+7psrTd/tCPH2qE7D2ZX3vro3sbazmyWOXOHfNx5PHLnHPlkp2ba5g/5vX43XbWed1ATASjtPtD+esjfBEv7mnng/vvY2vf+ReHmhax4ZyDx/ZextlRa4pe9QopVa+XKduYsDvGWNeFhEvcEZEXjDGXFjsD9q9uYKbw93s3lwx4/NyXafd0t7HD1/vpaHSM+6+zDx1KtXx2/u28uSxS7xzew2f/d4FRsNxWtr6GRgJU+CUJSuVTCl0Ct5CR3JjFvjj/TuslsfBqObZlVrFcjr9NcbcMMa8nPxzAHgd2JCLz6opLaDU46KmtGDG5y20S2Jqdp45u828b29jNQ2VHtpuWrn41H37GqvwBSPjXnfnhlIe3VHLkVdv8PqNANeHgvSPhIkZsNuW5szEIVDpceJx2gDhWFtf+mwklXbZ31zHgeb1027NOF1HSqXUyrBkF2NFZDOwEziRi/cfGYsRisQYGZu+e2W3L4Q/GGVvY+W8Lw5OVQufuZK1qcZLMBynyG3j+KVB/mjkVe69rQJB+OGFXtp7Rzi4p4G2ngAXuof5ztkunDYbDoHh8K0svH8s9xl5u8DWdUX86ls3saWqiCePXWJ0LE5b7/gL2jNdcNWqGqVWviUJ9CJSjNX58lPGmOEJjz0GPAZQXz//QHFtKMhYLMG1oeC0z2lp7+NYe/+U5Y/ZyqwiSQW5fY1V6RnvZ49c4MWLvdRXFBKMxDh5ZYDLfSN8cE89DVUe2ntG+Oz3LmBDGBgdY2QsTlmhEIkvURI+qdRtJwE8vN3a9anbF+LRHbWcuzZEx0Bw3AbjM6W5tKpGqZUv54FeRJxYQf6QMeY7Ex83xjwFPAXW5uDz/Zzt6728et3aWWk68w1KE/P6qZlr5vulAuGWKg+nrjq4d0sVnUOjnLvmZ2t1MQeaN3D+up9TVy7gcthoXOflui+IAYaWuId+oVP40NsaKC10jWucdqy9n72NlZR5XJMWdU01W9e+NEqtDjkN9CIiwN8CrxtjnsjlZ2FgNBybsdnLbDXfmYELGDerzQx4ZzuHOHSig4N7Gsa9X7cvxEg4TkWREwPsWF/CtcEQe26roK6skD/6zit0DYZwO4XhUISeQHRRvvpclRW6qa8sGjf2iQet6dJcmb8jTdsotTrkekZ/H/AbwKsici553x8aYxZ9G8KvH79KMJrg68ev8pl3b5/Xe2QGLmBcLj7z55ePXeLFi30EwzG+8KFddPtCHGnt5ty1IU5cGcIfjNDtG+M/vW0zH913W/p1w6EYcSAYNQSjSxvki13CmzeW07iumI3lnklnNRMPgkdau/nGyU52rPdyoPnW9fOJ1ygyf+oMX6mVKaeB3hjzEtYmJTlnJvycj72N1fiCEfzBKLu3VLCvMYovaPV0zwyCt1UV87NLA1y8GeAvnn+dwFiMb7/chTEGj8tJXZmbSBwu94/y0I4anvzxG5y4Mkjv8K3KFFngWOdqW00Jv/+uOzh0ooPdWypmDcQGQzAS53z38Lh+8hNn/pm/F53hK7Uy5U0LhEA4Me7nbKaafdaVFVLmcXG49QalHielHieHW2+kt/JLva6k0MnGikLae0b51ukuKopcBCPW54ZjEUIxGx6HnSv9o/zVv7dx9GI/E0e11DXyJYUuDp3o4IULPQQjMe5vWjcpRZUZ/A80b0CQSTtczZT+0guzSq1MeRPoa7wuegIRapIrSmcz3exzqmCV+efURcv7G9exqXyULVUeAmNxrvaNEElA3MBoOEEwnMDjdtDRH0sH+aWaxdsAhw2iCevzarwuPvFwI203A5y6MggGvtJyBX8wmj6YweTuk3NtDqedJ5VamfIm0O9trObwK91Zzyanm31mttfNvDD7paNWV8ndWyq4a2OQy/0jvGfnRs5dG+KN3gBOhxCOjA/jgbEYgbFbufilmsVXFDn56L7bKCt08a0z1/jtfVvZWV/OoRMd+EJRrg0FMcaMm63rLFyp/JU3gf7ljkHCMcPLHYNZPX+62Wfmdn4/vNDLqauDNK4r5psnO/GHYhx+5TpVxQW83DnEpb4RLveNEkuA035rH1WD9d9QMLrkKRoAXzDKcChGmcdFodPBYPI6w8E9DQA8uqOWwWBkyjy7Uir/5E2gvzYUGvdzvjIXQTVUeegYCFLjLcDtsDMyFuLizSjODTaKXQ4CoVg6sMfiLHse3mWHaBxixjrwlRQ6xjVX21lfPu3G6Voxo1T+yps2xetL3ON+ZppLP5ZUXxqD4dEdtTRUehCs3ZU8LgfV3gKiMUO3f4ybw2FcdquoaGnXtd7itlutDDaUunl4ew33bqmgsshJoctBS/sApR7nuMA93e9ioT2AlFIrV97M6AtdTiCc/GlJzVL9wSjH2q0+9dlsZZe6QNl6zceZjiHAOgAMjIS50j/KaMTKuydgiYpHp1ZZ5OLBO6o5uKeBk1cG0tcQ2noCU+7VCnO7CK2Uyg95M6P3hSLjfsKtoGYwU3ZfnK4T5bWBIC67MDQaZXA0Qk8gwmvX/SDgD8XwOB1UFTkpdtupKsquymexFdiFt2wspcbrpqakgDKPi2Pt/bT1BPjA7npqSqwunj3DY+O+497G6il/F7pBiFL5K29m9NtrS+gJ9LO9tiR931S9aDJNNbttae/jyKvd9I9EcNqE9aVuegJh+kfC9AyP4Q/FGA5F8bjtFDhs+Ja4Tw1YF333N9fR2uXn2mCQ+soimmq81JYO0lTjpdsX4rPfO8+F7mF21JUwHIqnv+Nc2kBo0FcqP+RNoD+aTM2kfsL89jNtqvHiLXDSH4gQisUptzupKnLT7Q9jl1u5+EA4TiC89Hu4uu3Cr91Tz3/cuYHgsctsqfKk+87c9Fubdrf1BLhwfZhgNMGWymLqKye3PJiOrm5VKv/kTaCfawuEVGOyR3fUcqS1G38oQmmhlYYpcFpb+V0dCHG1P4TdlqysWY5aSazLADaBkgIHLoeNvsAYp64MMjwWo76yiLqywkkHLV8wgiDsb66b08xcc/VK5Z+8CfSpVafZXhtNtQO40j/KcCiGLxihyO1gS6WH6mI3ocitFa3LUgyf4U3rvWyq8nB9aAyHXegbiUy67jDx7OVj998+r8/Sunql8k/eBPrpZvQTc86p24/uqAVgd0MFR9v6MMbwcqdVZeNxOwiMRdMHj8QyBnqnDfbfVcc9Wyr58rFLVBcXsKG8MD1T7/aFeOroJQyGe7ZUpittZpvFay5eqbUjbwK9Xaw+M/YJU/qW9j6ePdPFqauDPP7INg63XueZ0108emcNtd4C/vH4VcRmIxyN0xsI47KBzSaEIollncgXOoVyj4uGCk86gJ/vHiYU9fHhd2wG4OlTnfiDUZ45Y3XObO8d4aY/DMyeX9dcvFJrR94E+rgZ/xOsWasvGKG62EVHf5Ajrd20XvPjD0U5dLyTeCJBMJqgutiN1fnFmsEXuewML0M1DVj1rnYBl91OZZGb0WgivcnJnXUlXLgRQJB0oN7bWMn77t44aUY/m2xz8TrzV2r1y5tAP5XDrdf5+vEOilx2Ht5ei8HQNxImGk/gH4vhslsHhuGxKJGYlZGPJCAwtjw9agSw26G62EXvcIQr/SNsqS6mYyDIySsD3LWpnC1VRYBVHXSgeXLp6HQtDibKNhevM3+lVr+8DvTDoRj9gQg9iQTHCwb441/cQZnHxcnLg/zgwk0cNhvhkBXkbTYgWS05nGVP+4Vy2qDM46R/xDqw2ASKXA4KnA4Q6759jdXUV3rSq3trS920dg1T6nHOKfDOd2auVThKrX55HehLCh2UFzkJhuMMB2N8ueUSWyqLKXY7eNOGEkbG4ozFYoSiJh3kl5IAHqednZs8vH5zmLs2ltE/GuXtt1XSNxKmqthFaaEzHWRLPU6aarycujKILxih2xfKOmjPd2auVThKrX55G+hTS/43Vxbx+o1hbDZovebnWFs/0ViCcNzgsFl7KC2XSAJ6R8IggtNuR0QoKXTwjsYqHrmzlqdPdaaDc6nHmZ6Nt/UEJu18NRudmSu1duVloP/S0Tdovebjcl+Qm8NBgpE4dWUFFLrs/KS9n5HkFdvYEtZNSsbP0kIHd9R6ae8dochlB6Dc48QA7T0jPHnsEnduKB23+ClzNj6foK0zc6XWrrwM9M+c7mIsmmAkHGEknMBpFy73jrKjroTKYhf+sYX1rJ+rzIpPA3gLHHT7xxCB4bE4HpeNKm8Bb60vIxJPMBqO88QLF3n8kW3p3a7KPK5pF0cppdRM8qZ7Zaa3b62ivMiBDSGRMIxFDR1DIb5/vofOgaUN8qUFdtaXuvC4bNSUuKivKMRhF7qGQkTjCWw2AREaa4o5eO9m/vrX38qbNpTQ0R9M94Zf7M6Sc+nPr5Ra/fJyRn9taJTAWIyyIicjkTjRuEkvqFrqVa5jsTj+sThet513NFbz+CPbOH/dz5PHLvH+uzfhD0WT7Qw2pAP5449sG7dfbcpCa9ozt0lsaR8AtGRSqbUgLwP9WDSBLxgjnkgQSebj48twzdUOJJKVmm6HjYN7GtKB+tv/9b5pXzddamahNe2Z2yRO1ZNeKZWf8jLQY0xypry8ClxCZZGbG/4x1pcU8OWWS/QFrI1RUrn3uczQF1o5M1t/fqVUfsrLQH+2c2jZgzxAsdvJxnIPgbEYlwZG6RoK8raMzbrnOkNf6EVYvYir1NqUl4F+qfPwdsDpEOLJD7YB4QT0j0RImBESBuwi2Ox27tpYnp5Na227Umop5GXVTXRpOhikxQER63MN8NbNFdR4XXhcdmwCcWPYXOnhsX1b2N9cl37ddNU0WhWjlFpMeTmjX+q0jV3g9qpiOgaDFLoc/NZ9W7jSP8o3TnbidTkYDIb50L2b+cA92aVNtJGYUmox5WWgzzW3A4yBSNw6JSp2OdjZUMGmKg+vdPl57lwX77lrI5XFLi71jhCOJTja3guS3YVQTekopRaTBvp5CMcg2bmABBBLJPAW2NlYXs7pqz4uXB/G47qJDWFnfRmFTgdbKos43HoDfzA6rm/NVPSi6ey0T75S2dNAnyWnQDSZExKsnLzXbcdgcDsdlBa60vn3qbb16/aFqE8uVtK0zMJpekup7Gmgz1KBy06Fy4E/FAGEIredX9tdT0mhA0HSe7g+dv/W9Gum2gTkni2V4/rW6Mx0fjS9pVT2NNDPwCZWisYYaKwpprTARXtvgMYaL/feVjGubcFsUjPQA83jZ6A6M50fTW8plb2cBnoR+SqwH+g1xrwpl5+VC3fWlbCvsZrvvXqDaCzBQDDCvVsrefyRbbMG+Ikz9elmoNPdrzN9pdRiyfWM/u+B/w/4hxx/zqIpdNq4d0sFhW47H927lZqSAnoCY7zWNUyRUzi4pyGrwDtxpj7dDDRXfW2UUiolp4HeGHNMRDbn8jMWWyia4MW2fgD+9dWeSY+/5ws/pcAujE3okmYXcNisC7XRuFWNA/DTN/p54gcXCYZjxIzBBpR5XDTVeOkYHAWsvW13b65gZ315+oLu0Yu9vNY1xN7bKyfN7s92DnHoRAcH9zRQU1JAS3sfTTXe9MVfsDZGF4TdWyrGXRTONPF9M28DHGntntRZc6LFOPPQsxelcmvZc/Qi8hjwGEB9/eqYuU4M8mB1x4xPse+sAXqSjcxSRv1huv3hcQu7fnChhzf6Rin1OAH495/3Eo4ZvnGyk7Ii17jZ/aETHbxwwToI7d5cweHWG9SWDnLTH06/3zOnuxAR2noD6fsnnhlMPGvIvA3wzJkujDEzblm4GGceevaiVG4te6A3xjwFPAWwa9euldCLbEY2wJXljN6R7IFf7XVlPaNPzaYPvKWOn10e4OMP3D4pj39wT0P6Z01JAcCkGb0vGJk0o59o4vtO/OkPWr3yZ6psWYzqF62gUSq3xJjcxtZk6uZINhdjd+3aZU6fPp3T8SilVL4RkTPGmF3TPZ6XTc2UUkrdktNALyLfBH4GbBORLhH5cC4/Tyml1GS5rrr59Vy+v1JKqdlp6kYppfKcBnqllMpzGuiVUirPaaBXSqk8l/M6+rkQkT6gYwFvUQX0L9JwVpu1/N1hbX//tfzdQb9/FVBkjJl2xeGKCvQLJSKnZ1o0kM/W8neHtf391/J3B/3+2Xx/Td0opVSe00CvlFJ5Lt8C/VPLPYBltJa/O6zt77+Wvzvo95/1++dVjl4ppdRk+TajV0opNYEGeqWUynN5EehF5Ksi0isiry33WJaaiGwSkRdF5HUROS8in1zuMS0lESkQkZMi0pr8/v/Xco9pqYmIXUTOisiR5R7LUhORqyLyqoicE5E1tZmFiJSJyLMi8vPkv/+3TfvcfMjRi8g+YAT4h2w2OMknIrIeWG+MeVlEvMAZ4D8aYy4s89CWhIgI1mKRERFxAi8BnzTGHF/moS0ZEXkc2AWUGGP2L/d4lpKIXAV2GWPW3IIpEfka0GKM+YqIuACPMcY31XPzYkZvjDkGDC73OJaDMeaGMebl5J8DwOvAhuUd1dIxlpHkTWfyv9U/e8mSiGwEfhH4ynKPRS0dESkB9gF/C2CMiUwX5CFPAr2yJLdt3AmcWN6RLK1k6uIc0Au8YIxZS9//88DvY21TvBYZ4AcickZEHlvuwSyh24A+4O+SabuviEjRdE/WQJ8nRKQY+DbwKWPM8HKPZykZY+LGmLuAjcA9IrIm0ncish/oNcacWe6xLKP7jDFvBd4N/E4yjbsWOIC3Al80xuwERoH/c7ona6DPA8nc9LeBQ8aY7yz3eJZL8tT1x8C7lnkoS+U+4JeSeep/An5BRP5xeYe0tIwx3cmfvcBzwD3LO6Il0wV0ZZy9PosV+KekgX6VS16M/FvgdWPME8s9nqUmItUiUpb8cyHwMPDz5R3V0jDG/IExZqMxZjPwa8CPjDEfWuZhLRkRKUoWIJBMW7wTWBOVd8aYm8A1EdmWvOshYNoCjJzuGbtUkpuQPwBUiUgX8CfGmL9d3lEtmfuA3wBeTeapAf7QGPOvyzimpbQe+JqI2LEmLt8yxqy5MsM1qgZ4zprr4AC+YYx5fnmHtKT+G3AoWXFzGfgv0z0xL8orlVJKTU9TN0oplec00CulVJ7TQK+UUnlOA71SSuU5DfRKKZXnNNArpVSe00Cvlo2IjMz+rGlf+7si8oaIGBGpyrhfROSvko+9IiJTrhYUkVoR+ScRuSQiF0TkX0WkSUQeWIp2vyLy9yJyJdlet1VEHlrAe90hIj8TkbCIfHoxx6nygwZ6tVr9BGsVbMeE+98NNCb/ewz44sQXJlcTPwf82Biz1RizA/hDrAU4S+kzyR49nwKeXMD7DAKfAD63KKNSeUcDD5MLaQAAA1VJREFUvVp2yVn4X4jIa8lNJD6QvN8mIl9IbihyJDnrfi+AMeasMebqFG/3y1j7EphkT/qyZM/+TA8CUWNMOrgaY84ZY1qSN4szNnQ4lDwwICIPJTsFvprc7MYtIveIyHeSj/+yiIRExJXcEOVylr+Cn5HRWlpE7haRo8mOjN9PjV9EfiwinxeRnyZ/V/ckx95rjDkFRLP8PLXGaKBXK8GvAHcBzViz9L9IBrdfATYDbwY+Aky7g06GDcC1jNtdTO7P/yasDVqmsxNrlr0Dqx3sfSJSAPw98AFjzJuxltz/V+Dl5PMB9mL1WtkN7CH7dtHvAr4L6QZ1fw281xhzN/BV4H9lPLfIGPN24OPJx5SaVV70ulGr3juAbxpj4kCPiBzFCpbvAJ4xxiSAmyLyYhbvJVPcN9c+HyeNMV0Ayf5Bm4EAcMUY05Z8zteA3zHGfD55PWA7VufEJ7A2hLADLZPeeby/EJE/B9YB9ybv24Z1IHoheSJhB25kvOabYG22IyIlIlI204YTSoEGerUyTBWcZ7p/Jl3ApozbG4HuCc85D7x3hvcIZ/w5jvXvZKaxtGBdG4gCP8Sa+duB2S6Mfgb4DlZ+/WvA3cnPOW+Mme7sZeJBS5tVqVlp6katBMeADyR3iqrGmhGfxNr/9VeTufoarA6ls/kX4DeTef97Ab8x5saE5/wIcIvIR1N3iMhuEbl/hvf9ObBZRG5P3v4N4GjG+D8F/MwY0wdUAndgHVBmlDxb+UvAJiKPAheBaklu9CwiThG5M+MlqesX70h+N/9sn6GUBnq1EjwHvAK0YgXh30/22/421gz9NeBLWDlvP4CIfCLZknoj8IqIpPZM/Veslq1vAF/GymWPY6yWre8BHkmWV54H/pTJM//M14xhtYF9RkRexdq6L3Ux9wRWxc6x5O1XgFdMlq1hk8/7bPJ7R7DONv5MRFqBc8DbM54+JCI/TX72h5O/i9rk7+Jx4I9FpEusPUWVArRNsVrhRKTYGDMiIpVYs/z7kgeBNUdEfgx82hhzernHolYXzdGrle6IWDtIuYD/e60GeaUWQmf0SuWQiPwN1i5gmf7SGPN3yzEetTZpoFdKqTynF2OVUirPaaBXSqk8p4FeKaXynAZ6pZTKc/8/hwiaXLmS11EAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "##### Example code for how to compute correlations and draw scatter plots #####\n",
    "\n",
    "%pylab inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import scipy.stats\n",
    "OUTDIR=\"/datasets/cs185-sp21-A00-public/week4/precomputed/\"\n",
    "\n",
    "# Load data for each sample into a pandas data frame\n",
    "prefixes = [\"Chow_Rep1\",\"Chow_Rep2\",\"Chow_Rep3\"]\n",
    "data = pd.read_csv(os.path.join(OUTDIR, \"%s.genes.results\"%prefixes[0]), sep=\"\\t\")\n",
    "data = data[[\"gene_id\", \"TPM\"]]\n",
    "data.columns = [\"gene_id\", prefixes[0]]\n",
    "for i in range(1, len(prefixes)):\n",
    "    d = pd.read_csv(os.path.join(OUTDIR, \"%s.genes.results\"%prefixes[i]), sep=\"\\t\")\n",
    "    d = d[[\"gene_id\",\"TPM\"]]\n",
    "    d.columns = [\"gene_id\", prefixes[i]]\n",
    "    data = pd.merge(data, d, on=[\"gene_id\"])\n",
    "    \n",
    "# For each sample, take log10 TPM and add a pseudocount of 1 to avoid 0s\n",
    "for p in prefixes: data[p] = np.log10(data[p]+1)\n",
    "    \n",
    "# Correlation scatter plots\n",
    "fig = plt.figure()\n",
    "fig.set_size_inches((6, 4))\n",
    "ax = fig.add_subplot(111)\n",
    "ax.scatter((data[\"Chow_Rep1\"]+1), (data[\"Chow_Rep2\"]+1), s=2, alpha=0.5)\n",
    "ax.set_xlabel(\"log10 Chow_Rep1\")\n",
    "ax.set_ylabel(\"log10 Chow_Rep2\")\n",
    "\n",
    "\n",
    "# scipy.stats.pearsonr gives correlation (r, pval)\n",
    "print(scipy.stats.pearsonr((data[\"Chow_Rep2\"]+1), (data[\"Chow_Rep3\"]+1)))\n",
    "print(scipy.stats.pearsonr((data[\"Chow_Rep1\"]+1), (data[\"Chow_Rep2\"]+1)))\n",
    "\n",
    "# Pandas can give you pairwise correlation between all pairs of a set of columns\n",
    "# You can run a similar command in your report to directly print a nicely formatted correlation matrix table!\n",
    "(data[prefixes].corr())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "243e35c3159fc4c1610190c327f1ff17",
     "grade": false,
     "grade_id": "cell-f6a84764a7edc440",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "# 3. Differential expression analysis\n",
    "\n",
    "Now that we have quantified gene expression in each of our samples, we can perform differential expression analysis. For this, we will be using DESeq2, a widely used method that takes in \"counts\" (numbers of reads aligned to each gene) for different samples and outputs a set of differentially expressed genes.\n",
    "\n",
    "This tool is an R package, and so must be run from R. Fortunately, we can write R code directly in Jupyter notebooks using the R \"cell magic\" (see below). Alternatively, you can open a separate \"R\" document from JupyterHub or use R at the terminal if that's more comfortable. Since we have not run R before in this course, we have provided the necessary code to run DESeq2 below. The code does the following:\n",
    "\n",
    "* Loads the `DESeq2` library, as well as the `tximport` library which is needed to load RSEM results\n",
    "* Sets up the list of files to be used as input to DESeq2. We will input the RSEM results from each of our replicates in each condition.\n",
    "* Defines which condition each dataset belongs to (the first three are replicates of \"Chow\", the second three are replicates of \"HFD\").\n",
    "* Imports the data to DESeq2 using the `tximport` library\n",
    "* Runs DESeq2 to find differentially expressed genes.\n",
    "* Outputs results to `~/week4/chow_vs_hfd_deseq2.csv`.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q3)**</font> Summarize the methods used to detect differentially expressed genes. Even though you didn't have to write this code yourself, you still should summarize what was done following the prompts in the report document."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "code",
     "checksum": "76512568d045b2379b0c99adf45d04c8",
     "grade": false,
     "grade_id": "cell-8d0c80495750e889",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: It looks like you are importing RSEM genes.results files, setting txIn=FALSE\n",
      "\n",
      "R[write to console]: reading in files with read_tsv\n",
      "\n",
      "R[write to console]: 1 \n",
      "R[write to console]: 2 \n",
      "R[write to console]: 3 \n",
      "R[write to console]: 4 \n",
      "R[write to console]: 5 \n",
      "R[write to console]: 6 \n",
      "R[write to console]: using counts and average transcript lengths from tximport\n",
      "\n",
      "R[write to console]: estimating size factors\n",
      "\n",
      "R[write to console]: using 'avgTxLength' from assays(dds), correcting for library size\n",
      "\n",
      "R[write to console]: estimating dispersions\n",
      "\n",
      "R[write to console]: gene-wise dispersion estimates\n",
      "\n",
      "R[write to console]: mean-dispersion relationship\n",
      "\n",
      "R[write to console]: final dispersion estimates\n",
      "\n",
      "R[write to console]: fitting model and testing\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "##### Load the libraries we need #####\n",
    "library(\"DESeq2\")\n",
    "library(\"tximport\")\n",
    "\n",
    "##### List the files and set up metadata #####\n",
    "# Note, you should change this to use the files in your home directory\n",
    "files <- c(\"/datasets/cs185s-sp20-public/week4/precomputed/Chow_Rep1.genes.results\",\n",
    "         \"/datasets/cs185s-sp20-public/week4/precomputed/Chow_Rep2.genes.results\",\n",
    "         \"/datasets/cs185s-sp20-public/week4/precomputed/Chow_Rep3.genes.results\",\n",
    "         \"/datasets/cs185s-sp20-public/week4/precomputed/HFD_Rep1.genes.results\",\n",
    "         \"/datasets/cs185s-sp20-public/week4/precomputed/HFD_Rep2.genes.results\",\n",
    "         \"/datasets/cs185s-sp20-public/week4/precomputed/HFD_Rep3.genes.results\")\n",
    "conditions <- c(rep(\"Chow\", 3), rep(\"HFD\", 3))\n",
    "samples <- data.frame(\"run\"=c(\"Chow_Rep1\", \"Chow_Rep2\", \"Chow_Rep3\", \"HFD_Rep1\", \"HFD_Rep2\", \"HFD_Rep3\"),\n",
    "                    \"condition\"=conditions)\n",
    "names(files) = samples$run\n",
    "\n",
    "##### Use \"tximport\" to convert RSEM results to the format needed by DESeq2 #####\n",
    "txi <- tximport(files, type=\"rsem\")\n",
    "txi$length[txi$length == 0] <- 1 # add pseudocount of 1 to lengths to fix error with 0-length transcripts\n",
    "ddsTxi <- DESeqDataSetFromTximport(txi, colData=samples, design=~condition)\n",
    "\n",
    "# Filter things with very low counts to we don't waste time on those\n",
    "keep <- rowSums(counts(ddsTxi)) >= 10\n",
    "ddsTxi <- ddsTxi[keep,]\n",
    "\n",
    "##### Perform deseq2 #####\n",
    "dds <- DESeq(ddsTxi)\n",
    "res <- results(dds)\n",
    "\n",
    "##### Write results to chow_vs_hfd_deseq2.csv #####\n",
    "write.csv(as.data.frame(res), file=\"~/week4/chow_vs_hfd_deseq2.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "3c573bc170cf756a34f217057a471a77",
     "grade": false,
     "grade_id": "cell-fb50e02079c59383",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "Take a look at the results in the CSV file `~/week4/chow_vs_hfd_deseq2.csv`. The important columns to note are:\n",
    "\n",
    "* Column 1 (no column header): gives the gene id (e.g. ENSMUSG00000000088)\n",
    "* Column 3 (log2FoldChange): gives the log2 of the fold change in expression of the gene between conditions\n",
    "* Column 6 (\"pvalue\"): gives the p-value for each gene, where here the null hypothesis is that the gene is expressed equally in Chow vs. HFD.\n",
    "* Column 7 (\"padj\"): gives the pvalues \"adjusted\" for the number of hypotheses being tested (false discovery rate).\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q4)**</font> Following the prompts in the report notebook, visualize the differential expression results using a [\"volcano plot\"](https://en.wikipedia.org/wiki/Volcano_plot_(statistics)), which plots the log2 fold change vs. the -log10 p-value. We'll see an example in lecture this week of what this should look like.\n",
    "\n",
    "You'll notice all the gene ids are pretty cryptic looking (e.g. ENSMUSG00000000088). We have provided a file, `GRCm38.75.gene_names` in the `week4` directory that can be used to convert these to more canonical gene names. \n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q5)**</font>. At an adjusted p-value threshold (FDR) of 5%, how many genes were differentially expressed across the two conditions? Report the top 10 most differentially expressed genes, their log2 fold change, and the p-value. Report the gene names (e.g. *Fads2*) rather than gene ids."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "769d7c0314dd66cf5d6eaeeec8103a65",
     "grade": false,
     "grade_id": "cell-a3649e3bdd20eeea",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "# 4. Gene ontology enrichment analysis\n",
    "\n",
    "Finally, we will use gene ontology (GO) enrichment analysis to characterize the genes that were differentially expressed in Chow vs. HFD.\n",
    "\n",
    "We outline below a suggested way to do this using script `find_enrichment.py` from the python package `goatools`. However, there are many web tools available for performing GO analysis which you are welcome to use instead (although you're likely to get at least slightly different results with each, they should hopefully give back similar overall categories). For example these tools are pretty popular:\n",
    "\n",
    "* [DAVID](https://david.ncifcrf.gov/)\n",
    "* [PANTHER](http://www.pantherdb.org/)\n",
    "* [GORILLA](http://cbl-gorilla.cs.technion.ac.il/)\n",
    "\n",
    "For enrichment analysis, we need to define our \"foreground\" and \"background\". Here:\n",
    "* \"Background\" consists of all genes that we analyzed (i.e., all genes with any output in the DESeq2 csv file).\n",
    "* \"Foreground\" consists of all genes that were differentially expressed (padj<0.05). We will create separate foreground sets for up- and down-regulated genes.\n",
    "\n",
    "Then enrichment analysis will identify which annotations were over- or under-represented in our foreground vs. the background.\n",
    "\n",
    "The command `find_enrichment.py` can be used for GO analysis. Below is an example command:\n",
    "\n",
    "```shell\n",
    "find_enrichment.py sig_genes.txt all_genes.txt mgi.assoc \\\n",
    "    --obo go-basic.obo \\\n",
    "    --outfile=goea_fdr_bh.xlsx,goea_fdr_bh.tsv --method=fdr_bh --alpha=0.01\n",
    "```\n",
    "\n",
    "* `sig_genes.txt` is a file containing a list of gene names of differentially expressed genes.\n",
    "* `all_genes.txt` is a file containing a list of gene names of all analyzed genes.\n",
    "* `mgi.assoc` is a file mapping each gene to its annotated GO categories. We have provided this in the `week4/GOAnalysis/` folder\n",
    "* `go-basic.obo` is a file containing the GO database. It is also provided in the `week4/GOAnalysis/` folder.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Results-Q6)**</font>. Perform GO enrichment by comparing your differentially expressed genes (adjusted p<0.05) to all genes analyzed. **Perform a separate analaysis for down-regulated (log2 fold change <0) and up-regulated (log2fold change>0) genes**. For each enriched category, provide at least the name of the category and its enrichment p-value.\n",
    "\n",
    "<font color=\"red\">**FOR YOUR REPORT (Methods-Q4)**</font> Summarize the methods used to perform enrichment analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "f27945c61e271a6f2a46b4354dc0d464",
     "grade": false,
     "grade_id": "cell-da66487ba8b7859c",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "<font color=\"red\">**FOR YOUR REPORT (Discussion Q1-3)**</font>. Answer the discussion question prompts which will guide us through interpreting the results and considering how our analysis can be improved."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false,
    "nbgrader": {
     "cell_type": "markdown",
     "checksum": "104f8cc37cee97d4faa8563b5df69c04",
     "grade": false,
     "grade_id": "cell-b540071513547661",
     "locked": true,
     "schema_version": 3,
     "solution": false
    }
   },
   "source": [
    "## (Optional) Visualizing alignments in IGV\n",
    "\n",
    "It is a good practice to visualize our alignments, both to see that the alignment worked well, and also to see whether things we report as differentially expressed make sense. We'll use a genome browser called the [Integrative Genomics Viewer](http://software.broadinstitute.org/software/igv/), or IGV. \n",
    "\n",
    "Navigate to the web version of IGV: [https://igv.org/app/](https://igv.org/app/). Set the genome to GRCh38 (mm10). \n",
    "\n",
    "Take a moment to orient yourself with IGV. The top gives the names of each chromosome. The bottom track, labeled \"Refseq genes\" gives the names and coordinates for all annotated genes. Let's choose one to look at. Type \"Fads2\" in the search box at the top. This will zoom the view in on this gene. Notice how in the gene you can see the exon and intron structure. The little arrows in the introns point to the left, which means this gene is on the reverse strand of the reference. Take a look at another gene (e.g. Cyp2b10) to see a gene on the forward strand. Zoom in further (using the \"+\" at the top right) until you can see actual DNA sequence at the top.\n",
    "\n",
    "Now we'd like to load our sequence alignments. While IGV can directly visualize BAM files, we'll instead look at \"counts\" files (\".tdf\" format) which are much smaller. These give read counts per position (i.e. coverage) which can give us an idea of the abundance of each gene. We have preprocessed the BAM files into tdf using a command line tool offered by IGV (`igvtools`, which is not currently installed on datahub). To access these files locally on your desktop or laptop, first copy them to your home directory:\n",
    "\n",
    "```shell\n",
    "cp /datasets/cs185-sp21-A00-public/week4/tdfs/*.tdf .\n",
    "```\n",
    "\n",
    "Then from your JupyterHub home page you can click the checkbox next to each one and download the files. In IGV, click \"Tracks\" then \"Local file\" and select the tdf files you just downloaded.\n",
    "\n",
    "Navigate again to Fads2 (This should be one of your differentially expressed genes!). Note that the RNA-seq tracks have very \"spiky\" coverage. Some regions have tons of reads and others are flat. Note how that compares to the structure of the gene annotated on the bottom. As expected, the \"spikes\" correspond to reads from exons, since intron and intergenic sequences generally aren't sequenced in our RNA-seq experiment.\n",
    "\n",
    "Take note of the y-axis range for each dataset (after you initially load them, it should range from 0 to a couple hundred for this gene). To make the tracks more comparable, you can set them all to have the same data range by using the \"settings\" icon to the left of each track and clicking \"set data range\". Try setting them all to 500. (the appropriate range will be different for different genes, since genes are really highly expressed whereas others are very lowly expressed. See the huge range on your scatterplot!) \n",
    "\n",
    "It is aslo helpful to color tracks. For instance, you can color tracks by condition (see options in settings to the right of each track). An example screenshot is shown below.\n",
    "\n",
    "<img src=igv_screenshot_lab4.png width=1000>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# References\n",
    "\n",
    "To read more about the data used for this lab see [\"High fat diet-induced changes of mouse hepatic transcription and enhancer activity can be reversed by subsequent weight loss.\" Siersbaek et al. ](https://www.ncbi.nlm.nih.gov/pubmed/28071704)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
